// @wordpress/block-library/build/paragraph/edit.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _reactNative = require("react-native");

var _i18n = require("@wordpress/i18n");

var _blocks = require("@wordpress/blocks");

var _blockEditor = require("@wordpress/block-editor");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var name = 'core/paragraph';

var ParagraphEdit =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(ParagraphEdit, _Component);

  function ParagraphEdit(props) {
    var _this;

    (0, _classCallCheck2.default)(this, ParagraphEdit);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(ParagraphEdit).call(this, props));
    _this.onReplace = _this.onReplace.bind((0, _assertThisInitialized2.default)(_this));
    return _this;
  }

  (0, _createClass2.default)(ParagraphEdit, [{
    key: "onReplace",
    value: function onReplace(blocks) {
      var _this$props = this.props,
          attributes = _this$props.attributes,
          onReplace = _this$props.onReplace;
      onReplace(blocks.map(function (block, index) {
        return index === 0 && block.name === name ? (0, _objectSpread2.default)({}, block, {
          attributes: (0, _objectSpread2.default)({}, attributes, block.attributes)
        }) : block;
      }));
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          attributes = _this$props2.attributes,
          setAttributes = _this$props2.setAttributes,
          mergeBlocks = _this$props2.mergeBlocks,
          onReplace = _this$props2.onReplace,
          style = _this$props2.style;
      var placeholder = attributes.placeholder,
          content = attributes.content;
      return (0, _element.createElement)(_reactNative.View, null, (0, _element.createElement)(_blockEditor.RichText, {
        identifier: "content",
        tagName: "p",
        value: content,
        deleteEnter: true,
        style: style,
        onChange: function onChange(nextContent) {
          setAttributes({
            content: nextContent
          });
        },
        onSplit: function onSplit(value) {
          if (!value) {
            return (0, _blocks.createBlock)(name);
          }

          return (0, _blocks.createBlock)(name, (0, _objectSpread2.default)({}, attributes, {
            content: value
          }));
        },
        onMerge: mergeBlocks,
        onReplace: onReplace,
        onRemove: onReplace ? function () {
          return onReplace([]);
        } : undefined,
        placeholder: placeholder || (0, _i18n.__)('Start writing…')
      }));
    }
  }]);
  return ParagraphEdit;
}(_element.Component);

var _default = ParagraphEdit;
exports.default = _default;
