// @wordpress/block-library/build/verse/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.settings = exports.name = exports.metadata = void 0;

var _i18n = require("@wordpress/i18n");

var _deprecated = _interopRequireDefault(require("./deprecated"));

var _edit = _interopRequireDefault(require("./edit"));

var _icon = _interopRequireDefault(require("./icon"));

var _save = _interopRequireDefault(require("./save"));

var _tranforms = _interopRequireDefault(require("./tranforms"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var metadata = {
  name: "core/verse",
  category: "formatting",
  attributes: {
    content: {
      type: "string",
      source: "html",
      selector: "pre",
      "default": ""
    },
    textAlign: {
      type: "string"
    }
  }
};
exports.metadata = metadata;
var name = metadata.name;
exports.name = name;
var settings = {
  title: (0, _i18n.__)('Verse'),
  description: (0, _i18n.__)('Insert poetry. Use special spacing formats. Or quote song lyrics.'),
  icon: _icon.default,
  example: {
    attributes: {
      content: (0, _i18n.__)('WHAT was he doing, the great god Pan,') + '<br>' + (0, _i18n.__)('    Down in the reeds by the river?') + '<br>' + (0, _i18n.__)('Spreading ruin and scattering ban,') + '<br>' + (0, _i18n.__)('Splashing and paddling with hoofs of a goat,') + '<br>' + (0, _i18n.__)('And breaking the golden lilies afloat') + '<br>' + (0, _i18n.__)('    With the dragon-fly on the river.')
    }
  },
  keywords: [(0, _i18n.__)('poetry')],
  transforms: _tranforms.default,
  deprecated: _deprecated.default,
  merge: function merge(attributes, attributesToMerge) {
    return {
      content: attributes.content + attributesToMerge.content
    };
  },
  edit: _edit.default,
  save: _save.default
};
exports.settings = settings;
