// @wordpress/components/build-module/mobile/bottom-sheet/cell.native.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { TouchableOpacity, Text, View, TextInput, I18nManager } from 'react-native';
import { isEmpty } from 'lodash';
/**
 * WordPress dependencies
 */

import { Dashicon } from '@wordpress/components';
import { Component } from '@wordpress/element';
import { __, _x, sprintf } from '@wordpress/i18n';
/**
 * Internal dependencies
 */

import styles from './styles.scss';
import platformStyles from './cellStyles.scss';

var BottomSheetCell =
/*#__PURE__*/
function (_Component) {
  _inherits(BottomSheetCell, _Component);

  function BottomSheetCell(props) {
    var _this;

    _classCallCheck(this, BottomSheetCell);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(BottomSheetCell).apply(this, arguments));
    _this.state = {
      isEditingValue: props.autoFocus || false
    };
    return _this;
  }

  _createClass(BottomSheetCell, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate() {
      if (this.state.isEditingValue) {
        this._valueTextInput.focus();
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          accessibilityLabel = _this$props.accessibilityLabel,
          accessibilityHint = _this$props.accessibilityHint,
          accessibilityRole = _this$props.accessibilityRole,
          onPress = _this$props.onPress,
          label = _this$props.label,
          value = _this$props.value,
          _this$props$valuePlac = _this$props.valuePlaceholder,
          valuePlaceholder = _this$props$valuePlac === void 0 ? '' : _this$props$valuePlac,
          icon = _this$props.icon,
          _this$props$labelStyl = _this$props.labelStyle,
          labelStyle = _this$props$labelStyl === void 0 ? {} : _this$props$labelStyl,
          _this$props$valueStyl = _this$props.valueStyle,
          valueStyle = _this$props$valueStyl === void 0 ? {} : _this$props$valueStyl,
          onChangeValue = _this$props.onChangeValue,
          children = _this$props.children,
          _this$props$editable = _this$props.editable,
          editable = _this$props$editable === void 0 ? true : _this$props$editable,
          separatorType = _this$props.separatorType,
          _this$props$style = _this$props.style,
          style = _this$props$style === void 0 ? {} : _this$props$style,
          valueProps = _objectWithoutProperties(_this$props, ["accessibilityLabel", "accessibilityHint", "accessibilityRole", "onPress", "label", "value", "valuePlaceholder", "icon", "labelStyle", "valueStyle", "onChangeValue", "children", "editable", "separatorType", "style"]);

      var showValue = value !== undefined;
      var isValueEditable = editable && onChangeValue !== undefined;
      var defaultLabelStyle = showValue || icon !== undefined ? styles.cellLabel : styles.cellLabelCentered;
      var drawSeparator = separatorType && separatorType !== 'none' || separatorStyle === undefined;

      var onCellPress = function onCellPress() {
        if (isValueEditable) {
          startEditing();
        } else if (onPress !== undefined) {
          onPress();
        }
      };

      var finishEditing = function finishEditing() {
        _this2.setState({
          isEditingValue: false
        });
      };

      var startEditing = function startEditing() {
        if (_this2.state.isEditingValue === false) {
          _this2.setState({
            isEditingValue: true
          });
        }
      };

      var separatorStyle = function separatorStyle() {
        var leftMarginStyle = _objectSpread({}, styles.cellSeparator, platformStyles.separatorMarginLeft);

        switch (separatorType) {
          case 'leftMargin':
            return leftMarginStyle;

          case 'fullWidth':
            return styles.separator;

          case 'none':
            return undefined;

          case undefined:
            return showValue ? leftMarginStyle : styles.separator;
        }
      };

      var getValueComponent = function getValueComponent() {
        var styleRTL = I18nManager.isRTL && styles.cellValueRTL;

        var finalStyle = _objectSpread({}, styles.cellValue, valueStyle, styleRTL); // To be able to show the `middle` ellipsizeMode on editable cells
        // we show the TextInput just when the user wants to edit the value,
        // and the Text component to display it.
        // We also show the TextInput to display placeholder.


        var shouldShowPlaceholder = isValueEditable && value === '';
        return _this2.state.isEditingValue || shouldShowPlaceholder ? createElement(TextInput, _extends({
          ref: function ref(c) {
            return _this2._valueTextInput = c;
          },
          numberOfLines: 1,
          style: finalStyle,
          value: value,
          placeholder: valuePlaceholder,
          placeholderTextColor: '#87a6bc',
          onChangeText: onChangeValue,
          editable: isValueEditable,
          pointerEvents: _this2.state.isEditingValue ? 'auto' : 'none',
          onFocus: startEditing,
          onBlur: finishEditing
        }, valueProps)) : createElement(Text, {
          style: _objectSpread({}, styles.cellValue, valueStyle),
          numberOfLines: 1,
          ellipsizeMode: 'middle'
        }, value);
      };

      var getAccessibilityLabel = function getAccessibilityLabel() {
        if (accessibilityLabel || !showValue) {
          return accessibilityLabel || label;
        }

        return isEmpty(value) ? sprintf(
        /* translators: accessibility text. Empty state of a inline textinput cell. %s: The cell's title */
        _x('%s. Empty', 'inline textinput cell'), label) : // Separating by ',' is necessary to make a pause on urls (non-capitalized text)
        sprintf(
        /* translators: accessibility text. Inline textinput title and value.%1: Cell title, %2: cell value. */
        _x('%1$s, %2$s', 'inline textinput cell'), label, value);
      };

      return createElement(TouchableOpacity, {
        accessible: !this.state.isEditingValue,
        accessibilityLabel: getAccessibilityLabel(),
        accessibilityRole: accessibilityRole || 'button',
        accessibilityHint: isValueEditable ?
        /* translators: accessibility text */
        __('Double tap to edit this value') : accessibilityHint,
        onPress: onCellPress,
        style: _objectSpread({}, styles.clipToBounds, style)
      }, createElement(View, {
        style: styles.cellContainer
      }, createElement(View, {
        style: styles.cellRowContainer
      }, icon && createElement(View, {
        style: styles.cellRowContainer
      }, createElement(Dashicon, {
        icon: icon,
        size: 24
      }), createElement(View, {
        style: platformStyles.labelIconSeparator
      })), createElement(Text, {
        numberOfLines: 1,
        style: _objectSpread({}, defaultLabelStyle, labelStyle)
      }, label)), showValue && getValueComponent(), children), drawSeparator && createElement(View, {
        style: separatorStyle()
      }));
    }
  }]);

  return BottomSheetCell;
}(Component);

export { BottomSheetCell as default };
