// @wordpress/blocks/build/api/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "cloneBlock", {
  enumerable: true,
  get: function get() {
    return _factory.cloneBlock;
  }
});
Object.defineProperty(exports, "createBlock", {
  enumerable: true,
  get: function get() {
    return _factory.createBlock;
  }
});
Object.defineProperty(exports, "switchToBlockType", {
  enumerable: true,
  get: function get() {
    return _factory.switchToBlockType;
  }
});
Object.defineProperty(exports, "parse", {
  enumerable: true,
  get: function get() {
    return _parser.default;
  }
});
Object.defineProperty(exports, "getBlockAttributes", {
  enumerable: true,
  get: function get() {
    return _parser.getBlockAttributes;
  }
});
Object.defineProperty(exports, "parseWithAttributeSchema", {
  enumerable: true,
  get: function get() {
    return _parser.parseWithAttributeSchema;
  }
});
Object.defineProperty(exports, "serialize", {
  enumerable: true,
  get: function get() {
    return _serializer.default;
  }
});
Object.defineProperty(exports, "getBlockContent", {
  enumerable: true,
  get: function get() {
    return _serializer.getBlockContent;
  }
});
Object.defineProperty(exports, "getBlockDefaultClassName", {
  enumerable: true,
  get: function get() {
    return _serializer.getBlockDefaultClassName;
  }
});
Object.defineProperty(exports, "getSaveContent", {
  enumerable: true,
  get: function get() {
    return _serializer.getSaveContent;
  }
});
Object.defineProperty(exports, "registerBlockType", {
  enumerable: true,
  get: function get() {
    return _registration.registerBlockType;
  }
});
Object.defineProperty(exports, "unregisterBlockType", {
  enumerable: true,
  get: function get() {
    return _registration.unregisterBlockType;
  }
});
Object.defineProperty(exports, "getFreeformContentHandlerName", {
  enumerable: true,
  get: function get() {
    return _registration.getFreeformContentHandlerName;
  }
});
Object.defineProperty(exports, "setUnregisteredTypeHandlerName", {
  enumerable: true,
  get: function get() {
    return _registration.setUnregisteredTypeHandlerName;
  }
});
Object.defineProperty(exports, "getUnregisteredTypeHandlerName", {
  enumerable: true,
  get: function get() {
    return _registration.getUnregisteredTypeHandlerName;
  }
});
Object.defineProperty(exports, "getBlockType", {
  enumerable: true,
  get: function get() {
    return _registration.getBlockType;
  }
});
Object.defineProperty(exports, "getBlockTypes", {
  enumerable: true,
  get: function get() {
    return _registration.getBlockTypes;
  }
});
Object.defineProperty(exports, "getBlockSupport", {
  enumerable: true,
  get: function get() {
    return _registration.getBlockSupport;
  }
});
Object.defineProperty(exports, "hasBlockSupport", {
  enumerable: true,
  get: function get() {
    return _registration.hasBlockSupport;
  }
});
Object.defineProperty(exports, "isReusableBlock", {
  enumerable: true,
  get: function get() {
    return _registration.isReusableBlock;
  }
});
Object.defineProperty(exports, "getChildBlockNames", {
  enumerable: true,
  get: function get() {
    return _registration.getChildBlockNames;
  }
});
Object.defineProperty(exports, "hasChildBlocks", {
  enumerable: true,
  get: function get() {
    return _registration.hasChildBlocks;
  }
});
Object.defineProperty(exports, "hasChildBlocksWithInserterSupport", {
  enumerable: true,
  get: function get() {
    return _registration.hasChildBlocksWithInserterSupport;
  }
});
Object.defineProperty(exports, "setDefaultBlockName", {
  enumerable: true,
  get: function get() {
    return _registration.setDefaultBlockName;
  }
});
Object.defineProperty(exports, "getDefaultBlockName", {
  enumerable: true,
  get: function get() {
    return _registration.getDefaultBlockName;
  }
});
Object.defineProperty(exports, "isUnmodifiedDefaultBlock", {
  enumerable: true,
  get: function get() {
    return _utils.isUnmodifiedDefaultBlock;
  }
});
Object.defineProperty(exports, "normalizeIconObject", {
  enumerable: true,
  get: function get() {
    return _utils.normalizeIconObject;
  }
});
Object.defineProperty(exports, "pasteHandler", {
  enumerable: true,
  get: function get() {
    return _rawHandling.pasteHandler;
  }
});
Object.defineProperty(exports, "getPhrasingContentSchema", {
  enumerable: true,
  get: function get() {
    return _rawHandling.getPhrasingContentSchema;
  }
});
Object.defineProperty(exports, "children", {
  enumerable: true,
  get: function get() {
    return _children.default;
  }
});

var _factory = require("./factory");

var _parser = _interopRequireWildcard(require("./parser"));

var _serializer = _interopRequireWildcard(require("./serializer"));

var _registration = require("./registration");

var _utils = require("./utils");

var _rawHandling = require("./raw-handling");

var _children = _interopRequireDefault(require("./children"));
