// @wordpress/blocks/build/api/utils.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isUnmodifiedDefaultBlock = isUnmodifiedDefaultBlock;
exports.isValidIcon = isValidIcon;
exports.normalizeIconObject = normalizeIconObject;
exports.normalizeBlockType = normalizeBlockType;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _lodash = require("lodash");

var _tinycolor = _interopRequireWildcard(require("tinycolor2"));

var _element = require("@wordpress/element");

var _registration = require("./registration");

var _factory = require("./factory");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

/**
 * Array of icon colors containing a color to be used if the icon color
 * was not explicitly set but the icon background color was.
 *
 * @type {Object}
 */
var ICON_COLORS = ['#191e23', '#f8f9f9'];
/**
 * Determines whether the block is a default block
 * and its attributes are equal to the default attributes
 * which means the block is unmodified.
 *
 * @param  {WPBlock} block Block Object
 *
 * @return {boolean}       Whether the block is an unmodified default block
 */

function isUnmodifiedDefaultBlock(block) {
  var defaultBlockName = (0, _registration.getDefaultBlockName)();

  if (block.name !== defaultBlockName) {
    return false;
  } // Cache a created default block if no cache exists or the default block
  // name changed.


  if (!isUnmodifiedDefaultBlock.block || isUnmodifiedDefaultBlock.block.name !== defaultBlockName) {
    isUnmodifiedDefaultBlock.block = (0, _factory.createBlock)(defaultBlockName);
  }

  var newDefaultBlock = isUnmodifiedDefaultBlock.block;
  var blockType = (0, _registration.getBlockType)(defaultBlockName);
  return (0, _lodash.every)(blockType.attributes, function (value, key) {
    return newDefaultBlock.attributes[key] === block.attributes[key];
  });
}
/**
 * Function that checks if the parameter is a valid icon.
 *
 * @param {*} icon  Parameter to be checked.
 *
 * @return {boolean} True if the parameter is a valid icon and false otherwise.
 */


function isValidIcon(icon) {
  return !!icon && ((0, _lodash.isString)(icon) || (0, _element.isValidElement)(icon) || (0, _lodash.isFunction)(icon) || icon instanceof _element.Component);
}
/**
 * Function that receives an icon as set by the blocks during the registration
 * and returns a new icon object that is normalized so we can rely on just on possible icon structure
 * in the codebase.
 *
 * @param {WPBlockTypeIconRender} icon Render behavior of a block type icon;
 *                                     one of a Dashicon slug, an element, or a
 *                                     component.
 *
 * @return {WPBlockTypeIconDescriptor} Object describing the icon.
 */


function normalizeIconObject(icon) {
  if (isValidIcon(icon)) {
    return {
      src: icon
    };
  }

  if ((0, _lodash.has)(icon, ['background'])) {
    var tinyBgColor = (0, _tinycolor.default)(icon.background);
    return (0, _objectSpread2.default)({}, icon, {
      foreground: icon.foreground ? icon.foreground : (0, _tinycolor.mostReadable)(tinyBgColor, ICON_COLORS, {
        includeFallbackColors: true,
        level: 'AA',
        size: 'large'
      }).toHexString(),
      shadowColor: tinyBgColor.setAlpha(0.3).toRgbString()
    });
  }

  return icon;
}
/**
 * Normalizes block type passed as param. When string is passed then
 * it converts it to the matching block type object.
 * It passes the original object otherwise.
 *
 * @param {string|Object} blockTypeOrName  Block type or name.
 *
 * @return {?Object} Block type.
 */


function normalizeBlockType(blockTypeOrName) {
  if ((0, _lodash.isString)(blockTypeOrName)) {
    return (0, _registration.getBlockType)(blockTypeOrName);
  }

  return blockTypeOrName;
}
