// @wordpress/components/build/clipboard-button/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _clipboard = _interopRequireDefault(require("clipboard"));

var _classnames = _interopRequireDefault(require("classnames"));

var _iconButton = _interopRequireDefault(require("../icon-button"));

var _button = _interopRequireDefault(require("../button"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var ClipboardButton =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(ClipboardButton, _Component);

  function ClipboardButton() {
    var _this;

    (0, _classCallCheck2.default)(this, ClipboardButton);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(ClipboardButton).apply(this, arguments));
    _this.bindContainer = _this.bindContainer.bind((0, _assertThisInitialized2.default)(_this));
    _this.onCopy = _this.onCopy.bind((0, _assertThisInitialized2.default)(_this));
    _this.getText = _this.getText.bind((0, _assertThisInitialized2.default)(_this));
    return _this;
  }

  (0, _createClass2.default)(ClipboardButton, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var container = this.container,
          getText = this.getText,
          onCopy = this.onCopy;
      var button = container.firstChild;
      this.clipboard = new _clipboard.default(button, {
        text: getText,
        container: container
      });
      this.clipboard.on('success', onCopy);
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.clipboard.destroy();
      delete this.clipboard;
      clearTimeout(this.onCopyTimeout);
    }
  }, {
    key: "bindContainer",
    value: function bindContainer(container) {
      this.container = container;
    }
  }, {
    key: "onCopy",
    value: function onCopy(args) {
      // Clearing selection will move focus back to the triggering button,
      // ensuring that it is not reset to the body, and further that it is
      // kept within the rendered node.
      args.clearSelection();
      var _this$props = this.props,
          onCopy = _this$props.onCopy,
          onFinishCopy = _this$props.onFinishCopy;

      if (onCopy) {
        onCopy(); // For convenience and consistency, ClipboardButton offers to call
        // a secondary callback with delay. This is useful to reset
        // consumers' state, e.g. to revert a label from "Copied" to
        // "Copy".

        if (onFinishCopy) {
          clearTimeout(this.onCopyTimeout);
          this.onCopyTimeout = setTimeout(onFinishCopy, 4000);
        }
      }
    }
  }, {
    key: "getText",
    value: function getText() {
      var text = this.props.text;

      if ('function' === typeof text) {
        text = text();
      }

      return text;
    }
  }, {
    key: "render",
    value: function render() {
      // Disable reason: Exclude from spread props passed to Button
      // eslint-disable-next-line no-unused-vars
      var _this$props2 = this.props,
          className = _this$props2.className,
          children = _this$props2.children,
          onCopy = _this$props2.onCopy,
          onFinishCopy = _this$props2.onFinishCopy,
          text = _this$props2.text,
          buttonProps = (0, _objectWithoutProperties2.default)(_this$props2, ["className", "children", "onCopy", "onFinishCopy", "text"]);
      var icon = buttonProps.icon;
      var classes = (0, _classnames.default)('components-clipboard-button', className);
      var ComponentToUse = icon ? _iconButton.default : _button.default; // Workaround for inconsistent behavior in Safari, where <textarea> is not
      // the document.activeElement at the moment when the copy event fires.
      // This causes documentHasSelection() in the copy-handler component to
      // mistakenly override the ClipboardButton, and copy a serialized string
      // of the current block instead.

      var focusOnCopyEventTarget = function focusOnCopyEventTarget(event) {
        event.target.focus();
      };

      return (0, _element.createElement)("span", {
        ref: this.bindContainer,
        onCopy: focusOnCopyEventTarget
      }, (0, _element.createElement)(ComponentToUse, (0, _extends2.default)({}, buttonProps, {
        className: classes
      }), children));
    }
  }]);
  return ClipboardButton;
}(_element.Component);

var _default = ClipboardButton;
exports.default = _default;
