// @wordpress/block-library/build/heading/edit.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _headingToolbar = _interopRequireDefault(require("./heading-toolbar"));

var _editor = _interopRequireDefault(require("./editor.scss"));

var _reactNative = require("react-native");

var _i18n = require("@wordpress/i18n");

var _blockEditor = require("@wordpress/block-editor");

var _blocks = require("@wordpress/blocks");

/**
 * Internal dependencies
 */

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */
var HeadingEdit = function HeadingEdit(_ref) {
  var attributes = _ref.attributes,
      mergeBlocks = _ref.mergeBlocks,
      onFocus = _ref.onFocus,
      onReplace = _ref.onReplace,
      setAttributes = _ref.setAttributes,
      style = _ref.style;
  return (0, _element.createElement)(_reactNative.View, {
    onAccessibilityTap: onFocus
  }, (0, _element.createElement)(_blockEditor.BlockControls, null, (0, _element.createElement)(_headingToolbar.default, {
    minLevel: 2,
    maxLevel: 5,
    selectedLevel: attributes.level,
    onChange: function onChange(newLevel) {
      return setAttributes({
        level: newLevel
      });
    }
  })), (0, _element.createElement)(_blockEditor.RichText, {
    identifier: "content",
    tagName: 'h' + attributes.level,
    value: attributes.content,
    style: (0, _objectSpread2.default)({}, style, {
      minHeight: _editor.default['wp-block-heading'].minHeight
    }),
    onChange: function onChange(value) {
      return setAttributes({
        content: value
      });
    },
    onMerge: mergeBlocks,
    onSplit: function onSplit(value) {
      if (!value) {
        return (0, _blocks.createBlock)('core/paragraph');
      }

      return (0, _blocks.createBlock)('core/heading', (0, _objectSpread2.default)({}, attributes, {
        content: value
      }));
    },
    onReplace: onReplace,
    onRemove: function onRemove() {
      return onReplace([]);
    },
    placeholder: attributes.placeholder || (0, _i18n.__)('Write heading…')
  }));
};

var _default = HeadingEdit;
exports.default = _default;
