// @wordpress/block-library/build/heading/heading-toolbar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _lodash = require("lodash");

var _i18n = require("@wordpress/i18n");

var _components = require("@wordpress/components");

var _headingLevelIcon = _interopRequireDefault(require("./heading-level-icon"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var HeadingToolbar =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(HeadingToolbar, _Component);

  function HeadingToolbar() {
    (0, _classCallCheck2.default)(this, HeadingToolbar);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(HeadingToolbar).apply(this, arguments));
  }

  (0, _createClass2.default)(HeadingToolbar, [{
    key: "createLevelControl",
    value: function createLevelControl(targetLevel, selectedLevel, onChange) {
      return {
        icon: (0, _element.createElement)(_headingLevelIcon.default, {
          level: targetLevel
        }),
        // translators: %s: heading level e.g: "1", "2", "3"
        title: (0, _i18n.sprintf)((0, _i18n.__)('Heading %d'), targetLevel),
        isActive: targetLevel === selectedLevel,
        onClick: function onClick() {
          return onChange(targetLevel);
        }
      };
    }
  }, {
    key: "render",
    value: function render() {
      var _this = this;

      var _this$props = this.props,
          _this$props$isCollaps = _this$props.isCollapsed,
          isCollapsed = _this$props$isCollaps === void 0 ? true : _this$props$isCollaps,
          minLevel = _this$props.minLevel,
          maxLevel = _this$props.maxLevel,
          selectedLevel = _this$props.selectedLevel,
          onChange = _this$props.onChange;
      return (0, _element.createElement)(_components.Toolbar, {
        isCollapsed: isCollapsed,
        icon: (0, _element.createElement)(_headingLevelIcon.default, {
          level: selectedLevel
        }),
        controls: (0, _lodash.range)(minLevel, maxLevel).map(function (index) {
          return _this.createLevelControl(index, selectedLevel, onChange);
        })
      });
    }
  }]);
  return HeadingToolbar;
}(_element.Component);

var _default = HeadingToolbar;
exports.default = _default;
