// @wordpress/block-library/build/heading/transforms.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _blocks = require("@wordpress/blocks");

var _shared = require("./shared");

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var transforms = {
  from: [{
    type: 'block',
    blocks: ['core/paragraph'],
    transform: function transform(_ref) {
      var content = _ref.content;
      return (0, _blocks.createBlock)('core/heading', {
        content: content
      });
    }
  }, {
    type: 'raw',
    selector: 'h1,h2,h3,h4,h5,h6',
    schema: {
      h1: {
        children: (0, _blocks.getPhrasingContentSchema)()
      },
      h2: {
        children: (0, _blocks.getPhrasingContentSchema)()
      },
      h3: {
        children: (0, _blocks.getPhrasingContentSchema)()
      },
      h4: {
        children: (0, _blocks.getPhrasingContentSchema)()
      },
      h5: {
        children: (0, _blocks.getPhrasingContentSchema)()
      },
      h6: {
        children: (0, _blocks.getPhrasingContentSchema)()
      }
    },
    transform: function transform(node) {
      return (0, _blocks.createBlock)('core/heading', (0, _objectSpread2.default)({}, (0, _blocks.getBlockAttributes)('core/heading', node.outerHTML), {
        level: (0, _shared.getLevelFromHeadingNodeName)(node.nodeName)
      }));
    }
  }].concat((0, _toConsumableArray2.default)([2, 3, 4, 5, 6].map(function (level) {
    return {
      type: 'prefix',
      prefix: Array(level + 1).join('#'),
      transform: function transform(content) {
        return (0, _blocks.createBlock)('core/heading', {
          level: level,
          content: content
        });
      }
    };
  }))),
  to: [{
    type: 'block',
    blocks: ['core/paragraph'],
    transform: function transform(_ref2) {
      var content = _ref2.content;
      return (0, _blocks.createBlock)('core/paragraph', {
        content: content
      });
    }
  }]
};
var _default = transforms;
exports.default = _default;
