// @wordpress/block-editor/build/components/caption/index.native.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _reactNative = require("react-native");

var _blockEditor = require("@wordpress/block-editor");

var _compose = require("@wordpress/compose");

var _data = require("@wordpress/data");

var _i18n = require("@wordpress/i18n");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */
var Caption = function Caption(_ref) {
  var accessible = _ref.accessible,
      accessibilityLabel = _ref.accessibilityLabel,
      onBlur = _ref.onBlur,
      onChange = _ref.onChange,
      onFocus = _ref.onFocus,
      isSelected = _ref.isSelected,
      shouldDisplay = _ref.shouldDisplay,
      text = _ref.text;
  return (0, _element.createElement)(_reactNative.View, {
    style: {
      padding: 12,
      flex: 1,
      display: shouldDisplay ? 'flex' : 'none'
    },
    accessible: accessible,
    accessibilityLabel: accessibilityLabel,
    accessibilityRole: 'button'
  }, (0, _element.createElement)(_blockEditor.RichText, {
    rootTagsToEliminate: ['p'],
    placeholder: (0, _i18n.__)('Write caption…'),
    value: text,
    onChange: onChange,
    unstableOnFocus: onFocus,
    onBlur: onBlur // always assign onBlur as props
    ,
    isSelected: isSelected,
    __unstableMobileNoFocusOnMount: true,
    fontSize: 14,
    underlineColorAndroid: "transparent",
    textAlign: 'center',
    tagName: ''
  }));
};

var _default = (0, _compose.compose)([(0, _data.withSelect)(function (select, _ref2) {
  var accessibilityLabelCreator = _ref2.accessibilityLabelCreator,
      clientId = _ref2.clientId;

  var _select = select('core/block-editor'),
      getBlockAttributes = _select.getBlockAttributes,
      getSelectedBlockClientId = _select.getSelectedBlockClientId;

  var _getBlockAttributes = getBlockAttributes(clientId),
      caption = _getBlockAttributes.caption;

  var accessibilityLabel = accessibilityLabelCreator ? accessibilityLabelCreator(caption) : undefined;
  var isBlockSelected = getSelectedBlockClientId() === clientId; // We'll render the caption so that the soft keyboard is not forced to close on Android
  // but still hide it by setting its display style to none. See wordpress-mobile/gutenberg-mobile#1221

  var shouldDisplay = !_blockEditor.RichText.isEmpty(caption) > 0 || isBlockSelected;
  return {
    accessibilityLabel: accessibilityLabel,
    shouldDisplay: shouldDisplay,
    text: caption
  };
}), (0, _data.withDispatch)(function (dispatch, _ref3) {
  var clientId = _ref3.clientId;

  var _dispatch = dispatch('core/block-editor'),
      updateBlockAttributes = _dispatch.updateBlockAttributes;

  return {
    onChange: function onChange(caption) {
      updateBlockAttributes(clientId, {
        caption: caption
      });
    }
  };
})])(Caption);

exports.default = _default;
