// @wordpress/block-library/build/preformatted/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.settings = exports.name = exports.metadata = void 0;

var _i18n = require("@wordpress/i18n");

var _edit = _interopRequireDefault(require("./edit"));

var _icon = _interopRequireDefault(require("./icon"));

var _save = _interopRequireDefault(require("./save"));

var _transforms = _interopRequireDefault(require("./transforms"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var metadata = {
  name: "core/preformatted",
  category: "formatting",
  attributes: {
    content: {
      type: "string",
      source: "html",
      selector: "pre",
      "default": ""
    }
  }
};
exports.metadata = metadata;
var name = metadata.name;
exports.name = name;
var settings = {
  title: (0, _i18n.__)('Preformatted'),
  description: (0, _i18n.__)('Add text that respects your spacing and tabs, and also allows styling.'),
  icon: _icon.default,
  example: {
    attributes: {
      content: (0, _i18n.__)('EXT. XANADU - FAINT DAWN - 1940 (MINIATURE)') + '\n' + (0, _i18n.__)('Window, very small in the distance, illuminated.') + '\n' + (0, _i18n.__)('All around this is an almost totally black screen. Now, as the camera moves slowly towards the window which is almost a postage stamp in the frame, other forms appear;')
    }
  },
  transforms: _transforms.default,
  edit: _edit.default,
  save: _save.default,
  merge: function merge(attributes, attributesToMerge) {
    return {
      content: attributes.content + attributesToMerge.content
    };
  }
};
exports.settings = settings;
