// @wordpress/block-editor/build/components/inner-blocks/template-picker.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _classnames = _interopRequireDefault(require("classnames"));

var _i18n = require("@wordpress/i18n");

var _components = require("@wordpress/components");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */
function InnerBlocksTemplatePicker(_ref) {
  var options = _ref.options,
      onSelect = _ref.onSelect,
      allowSkip = _ref.allowSkip;
  var classes = (0, _classnames.default)('block-editor-inner-blocks__template-picker', {
    'has-many-options': options.length > 4
  });
  var instructions = allowSkip ? (0, _i18n.__)('Select a layout to start with, or make one yourself.') : (0, _i18n.__)('Select a layout to start with.');
  return (0, _element.createElement)(_components.Placeholder, {
    icon: "layout",
    label: (0, _i18n.__)('Choose Layout'),
    instructions: instructions,
    className: classes
  }, (0, _element.createElement)("ul", {
    className: "block-editor-inner-blocks__template-picker-options",
    role: "list"
  }, options.map(function (templateOption, index) {
    return (0, _element.createElement)("li", {
      key: index
    }, (0, _element.createElement)(_components.IconButton, {
      isLarge: true,
      icon: templateOption.icon,
      onClick: function onClick() {
        return onSelect(templateOption.template);
      },
      className: "block-editor-inner-blocks__template-picker-option",
      label: templateOption.title
    }));
  })), allowSkip && (0, _element.createElement)("div", {
    className: "block-editor-inner-blocks__template-picker-skip"
  }, (0, _element.createElement)(_components.Button, {
    isLink: true,
    onClick: function onClick() {
      return onSelect(undefined);
    }
  }, (0, _i18n.__)('Skip'))));
}

var _default = InnerBlocksTemplatePicker;
exports.default = _default;
