// @wordpress/components/build-module/mobile/html-text-input/index.native.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _assertThisInitialized from "@babel/runtime/helpers/esm/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { TextInput } from 'react-native';
/**
 * WordPress dependencies
 */

import { Component } from '@wordpress/element';
import { __ } from '@wordpress/i18n';
import { parse } from '@wordpress/blocks';
import { withDispatch, withSelect } from '@wordpress/data';
import { withInstanceId, compose } from '@wordpress/compose';
/**
 * Internal dependencies
 */

import HTMLInputContainer from './container';
import styles from './style.scss';
export var HTMLTextInput =
/*#__PURE__*/
function (_Component) {
  _inherits(HTMLTextInput, _Component);

  function HTMLTextInput() {
    var _this;

    _classCallCheck(this, HTMLTextInput);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(HTMLTextInput).apply(this, arguments));
    _this.edit = _this.edit.bind(_assertThisInitialized(_this));
    _this.stopEditing = _this.stopEditing.bind(_assertThisInitialized(_this));
    _this.state = {
      isDirty: false,
      value: ''
    };
    return _this;
  }

  _createClass(HTMLTextInput, [{
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      //TODO: Blocking main thread
      this.stopEditing();
    }
  }, {
    key: "edit",
    value: function edit(html) {
      this.props.onChange(html);
      this.setState({
        value: html,
        isDirty: true
      });
    }
  }, {
    key: "stopEditing",
    value: function stopEditing() {
      if (this.state.isDirty) {
        this.props.onPersist(this.state.value);
        this.setState({
          isDirty: false
        });
      }
    }
  }, {
    key: "render",
    value: function render() {
      return createElement(HTMLInputContainer, {
        parentHeight: this.props.parentHeight
      }, createElement(TextInput, {
        autoCorrect: false,
        accessibilityLabel: "html-view-title",
        textAlignVertical: "center",
        numberOfLines: 1,
        style: styles.htmlViewTitle,
        value: this.props.title,
        placeholder: __('Add title'),
        onChangeText: this.props.editTitle
      }), createElement(TextInput, {
        autoCorrect: false,
        accessibilityLabel: "html-view-content",
        textAlignVertical: "top",
        multiline: true,
        style: styles.htmlView,
        value: this.state.value,
        onChangeText: this.edit,
        onBlur: this.stopEditing,
        placeholder: __('Start writing…'),
        scrollEnabled: HTMLInputContainer.scrollEnabled
      }));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(props, state) {
      if (state.isDirty) {
        return null;
      }

      return {
        value: props.value,
        isDirty: false
      };
    }
  }]);

  return HTMLTextInput;
}(Component);
export default compose([withSelect(function (select) {
  var _select = select('core/editor'),
      getEditedPostAttribute = _select.getEditedPostAttribute,
      getEditedPostContent = _select.getEditedPostContent;

  return {
    title: getEditedPostAttribute('title'),
    value: getEditedPostContent()
  };
}), withDispatch(function (dispatch) {
  var _dispatch = dispatch('core/block-editor'),
      resetBlocks = _dispatch.resetBlocks;

  var _dispatch2 = dispatch('core/editor'),
      editPost = _dispatch2.editPost;

  return {
    editTitle: function editTitle(title) {
      editPost({
        title: title
      });
    },
    onChange: function onChange(content) {
      editPost({
        content: content
      });
    },
    onPersist: function onPersist(content) {
      resetBlocks(parse(content));
    }
  };
}), withInstanceId])(HTMLTextInput);
