// @wordpress/components/build/form-file-upload/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _iconButton = _interopRequireDefault(require("../icon-button"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var FormFileUpload =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(FormFileUpload, _Component);

  function FormFileUpload() {
    var _this;

    (0, _classCallCheck2.default)(this, FormFileUpload);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(FormFileUpload).apply(this, arguments));
    _this.openFileDialog = _this.openFileDialog.bind((0, _assertThisInitialized2.default)(_this));
    _this.bindInput = _this.bindInput.bind((0, _assertThisInitialized2.default)(_this));
    return _this;
  }

  (0, _createClass2.default)(FormFileUpload, [{
    key: "openFileDialog",
    value: function openFileDialog() {
      this.input.click();
    }
  }, {
    key: "bindInput",
    value: function bindInput(ref) {
      this.input = ref;
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          accept = _this$props.accept,
          children = _this$props.children,
          _this$props$icon = _this$props.icon,
          icon = _this$props$icon === void 0 ? 'upload' : _this$props$icon,
          _this$props$multiple = _this$props.multiple,
          multiple = _this$props$multiple === void 0 ? false : _this$props$multiple,
          onChange = _this$props.onChange,
          render = _this$props.render,
          props = (0, _objectWithoutProperties2.default)(_this$props, ["accept", "children", "icon", "multiple", "onChange", "render"]);
      var ui = render ? render({
        openFileDialog: this.openFileDialog
      }) : (0, _element.createElement)(_iconButton.default, (0, _extends2.default)({
        icon: icon,
        onClick: this.openFileDialog
      }, props), children);
      return (0, _element.createElement)("div", {
        className: "components-form-file-upload"
      }, ui, (0, _element.createElement)("input", {
        type: "file",
        ref: this.bindInput,
        multiple: multiple,
        style: {
          display: 'none'
        },
        accept: accept,
        onChange: onChange
      }));
    }
  }]);
  return FormFileUpload;
}(_element.Component);

var _default = FormFileUpload;
exports.default = _default;
