// @wordpress/components/build-module/button/index.native.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread";
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { StyleSheet, TouchableOpacity, Text, View, Platform } from 'react-native';
var isAndroid = Platform.OS === 'android';
var marginBottom = isAndroid ? -0.5 : 0;
var marginLeft = -3;
var styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 3,
    justifyContent: 'center',
    alignItems: 'center'
  },
  buttonInactive: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    aspectRatio: 1,
    backgroundColor: 'white'
  },
  buttonActive: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    borderRadius: 6,
    borderColor: '#2e4453',
    aspectRatio: 1,
    backgroundColor: '#2e4453'
  },
  subscriptInactive: {
    color: '#7b9ab1',
    fontWeight: 'bold',
    fontSize: 13,
    alignSelf: 'flex-end',
    marginLeft: marginLeft,
    marginBottom: marginBottom
  },
  subscriptActive: {
    color: 'white',
    fontWeight: 'bold',
    fontSize: 13,
    alignSelf: 'flex-end',
    marginLeft: marginLeft,
    marginBottom: marginBottom
  }
});
export default function Button(props) {
  var children = props.children,
      onClick = props.onClick,
      disabled = props.disabled,
      hint = props.hint,
      ariaDisabled = props['aria-disabled'],
      ariaLabel = props['aria-label'],
      ariaPressed = props['aria-pressed'],
      subscript = props['data-subscript'];
  var isDisabled = ariaDisabled || disabled;

  var buttonViewStyle = _objectSpread({
    opacity: isDisabled ? 0.2 : 1
  }, ariaPressed ? styles.buttonActive : styles.buttonInactive);

  var states = [];

  if (ariaPressed) {
    states.push('selected');
  }

  if (isDisabled) {
    states.push('disabled');
  }

  return createElement(TouchableOpacity, {
    activeOpacity: 0.7,
    accessible: true,
    accessibilityLabel: ariaLabel,
    accessibilityStates: states,
    accessibilityRole: 'button',
    accessibilityHint: hint,
    onPress: onClick,
    style: styles.container,
    disabled: isDisabled
  }, createElement(View, {
    style: buttonViewStyle
  }, createElement(View, {
    style: {
      flexDirection: 'row'
    }
  }, children, subscript && createElement(Text, {
    style: ariaPressed ? styles.subscriptActive : styles.subscriptInactive
  }, subscript))));
}
