// @wordpress/block-library/build-module/separator/edit.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import { createElement, Fragment } from "@wordpress/element";

/**
 * External dependencies
 */
import classnames from 'classnames';
/**
 * WordPress dependencies
 */

import { __ } from '@wordpress/i18n';
import { HorizontalRule } from '@wordpress/components';
import { InspectorControls, withColors, PanelColorSettings } from '@wordpress/block-editor';

function SeparatorEdit(_ref) {
  var color = _ref.color,
      setColor = _ref.setColor,
      className = _ref.className;
  return createElement(Fragment, null, createElement(HorizontalRule, {
    className: classnames(className, _defineProperty({
      'has-background': color.color
    }, color.class, color.class)),
    style: {
      backgroundColor: color.color,
      color: color.color
    }
  }), createElement(InspectorControls, null, createElement(PanelColorSettings, {
    title: __('Color Settings'),
    colorSettings: [{
      value: color.color,
      onChange: setColor,
      label: __('Color')
    }]
  })));
}

export default withColors('color', {
  textColor: 'color'
})(SeparatorEdit);
