// @wordpress/block-library/build/navigation-menu-item/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.settings = exports.name = exports.metadata = void 0;

var _i18n = require("@wordpress/i18n");

var _edit = _interopRequireDefault(require("./edit"));

var _save = _interopRequireDefault(require("./save"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var metadata = {
  name: "core/navigation-menu-item",
  category: "layout",
  attributes: {
    label: {
      type: "string"
    },
    destination: {
      type: "string"
    },
    nofollow: {
      type: "boolean",
      "default": false
    },
    title: {
      type: "string"
    },
    description: {
      type: "string"
    },
    opensInNewTab: {
      type: "boolean",
      "default": false
    }
  }
};
exports.metadata = metadata;
var name = metadata.name;
exports.name = name;
var settings = {
  title: (0, _i18n.__)('Menu Item (Experimental)'),
  parent: ['core/navigation-menu'],
  icon: 'admin-links',
  description: (0, _i18n.__)('Add a page, link, or other item to your Navigation Menu.'),
  edit: _edit.default,
  save: _save.default
};
exports.settings = settings;
