// @wordpress/block-library/build/navigation-menu-item/menu-item-actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _data = require("@wordpress/data");

var _compose = require("@wordpress/compose");

/**
 * WordPress dependencies
 */
function MenuItemActions(_ref) {
  var destination = _ref.destination,
      moveLeft = _ref.moveLeft,
      moveRight = _ref.moveRight,
      moveToEnd = _ref.moveToEnd,
      moveToStart = _ref.moveToStart,
      onEditLableClicked = _ref.onEditLableClicked,
      remove = _ref.remove;
  return (0, _element.createElement)(_components.NavigableMenu, null, (0, _element.createElement)(_components.MenuItem, {
    icon: "admin-links"
  }, destination), (0, _element.createElement)(_components.MenuItem, {
    onClick: onEditLableClicked,
    icon: "edit"
  }, (0, _i18n.__)('Edit label text')), (0, _element.createElement)("div", {
    className: "wp-block-navigation-menu-item__separator"
  }), (0, _element.createElement)(_components.MenuItem, {
    onClick: moveToStart,
    icon: "arrow-up-alt2"
  }, (0, _i18n.__)('Move to start')), (0, _element.createElement)(_components.MenuItem, {
    onClick: moveLeft,
    icon: "arrow-left-alt2"
  }, (0, _i18n.__)('Move left')), (0, _element.createElement)(_components.MenuItem, {
    onClick: moveRight,
    icon: "arrow-right-alt2"
  }, (0, _i18n.__)('Move right')), (0, _element.createElement)(_components.MenuItem, {
    onClick: moveToEnd,
    icon: "arrow-down-alt2"
  }, (0, _i18n.__)('Move to end')), (0, _element.createElement)(_components.MenuItem, {
    icon: "arrow-left-alt2"
  }, (0, _i18n.__)('Nest underneath…')), (0, _element.createElement)("div", {
    className: "navigation-menu-item__separator"
  }), (0, _element.createElement)(_components.MenuItem, {
    onClick: remove,
    icon: "trash"
  }, (0, _i18n.__)('Remove from menu')));
}

var _default = (0, _compose.compose)([(0, _data.withDispatch)(function (dispatch, _ref2, _ref3) {
  var clientId = _ref2.clientId;
  var select = _ref3.select;

  var _select = select('core/block-editor'),
      getBlockOrder = _select.getBlockOrder,
      getBlockRootClientId = _select.getBlockRootClientId;

  var parentID = getBlockRootClientId(clientId);

  var _dispatch = dispatch('core/block-editor'),
      moveBlocksDown = _dispatch.moveBlocksDown,
      moveBlocksUp = _dispatch.moveBlocksUp,
      moveBlockToPosition = _dispatch.moveBlockToPosition,
      removeBlocks = _dispatch.removeBlocks;

  return {
    moveToStart: function moveToStart() {
      moveBlockToPosition(clientId, parentID, parentID, 0);
    },
    moveRight: function moveRight() {
      moveBlocksDown(clientId, parentID);
    },
    moveLeft: function moveLeft() {
      moveBlocksUp(clientId, parentID);
    },
    moveToEnd: function moveToEnd() {
      moveBlockToPosition(clientId, parentID, parentID, getBlockOrder(parentID).length - 1);
    },
    remove: function remove() {
      removeBlocks(clientId);
    }
  };
})])(MenuItemActions);

exports.default = _default;
