// @wordpress/components/build/button/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Button;

var _element = require("@wordpress/element");

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _reactNative = require("react-native");

/**
 * External dependencies
 */
var isAndroid = _reactNative.Platform.OS === 'android';
var marginBottom = isAndroid ? -0.5 : 0;
var marginLeft = -3;

var styles = _reactNative.StyleSheet.create({
  container: {
    flex: 1,
    padding: 3,
    justifyContent: 'center',
    alignItems: 'center'
  },
  buttonInactive: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    aspectRatio: 1,
    backgroundColor: 'white'
  },
  buttonActive: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    borderRadius: 6,
    borderColor: '#2e4453',
    aspectRatio: 1,
    backgroundColor: '#2e4453'
  },
  subscriptInactive: {
    color: '#7b9ab1',
    fontWeight: 'bold',
    fontSize: 13,
    alignSelf: 'flex-end',
    marginLeft: marginLeft,
    marginBottom: marginBottom
  },
  subscriptActive: {
    color: 'white',
    fontWeight: 'bold',
    fontSize: 13,
    alignSelf: 'flex-end',
    marginLeft: marginLeft,
    marginBottom: marginBottom
  }
});

function Button(props) {
  var children = props.children,
      onClick = props.onClick,
      disabled = props.disabled,
      hint = props.hint,
      ariaDisabled = props['aria-disabled'],
      ariaLabel = props['aria-label'],
      ariaPressed = props['aria-pressed'],
      subscript = props['data-subscript'];
  var isDisabled = ariaDisabled || disabled;
  var buttonViewStyle = (0, _objectSpread2.default)({
    opacity: isDisabled ? 0.2 : 1
  }, ariaPressed ? styles.buttonActive : styles.buttonInactive);
  var states = [];

  if (ariaPressed) {
    states.push('selected');
  }

  if (isDisabled) {
    states.push('disabled');
  }

  return (0, _element.createElement)(_reactNative.TouchableOpacity, {
    activeOpacity: 0.7,
    accessible: true,
    accessibilityLabel: ariaLabel,
    accessibilityStates: states,
    accessibilityRole: 'button',
    accessibilityHint: hint,
    onPress: onClick,
    style: styles.container,
    disabled: isDisabled
  }, (0, _element.createElement)(_reactNative.View, {
    style: buttonViewStyle
  }, (0, _element.createElement)(_reactNative.View, {
    style: {
      flexDirection: 'row'
    }
  }, children, subscript && (0, _element.createElement)(_reactNative.Text, {
    style: ariaPressed ? styles.subscriptActive : styles.subscriptInactive
  }, subscript))));
}
