// @wordpress/components/build/mobile/bottom-sheet/button.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _reactNative = require("react-native");

var _styles = _interopRequireDefault(require("./styles.scss"));

/**
 * External dependencies
 */

/**
 * Internal dependencies
 */
var BottomSheetButton = function BottomSheetButton(_ref) {
  var onPress = _ref.onPress,
      disabled = _ref.disabled,
      text = _ref.text,
      color = _ref.color;
  return (0, _element.createElement)(_reactNative.TouchableOpacity, {
    accessible: true,
    onPress: onPress,
    disabled: disabled
  }, (0, _element.createElement)(_reactNative.View, {
    style: {
      flexDirection: 'row',
      justifyContent: 'center'
    }
  }, (0, _element.createElement)(_reactNative.Text, {
    style: (0, _objectSpread2.default)({}, _styles.default.buttonText, {
      color: color
    })
  }, text)));
};

var _default = BottomSheetButton;
exports.default = _default;
