// @angular-react/core/esm5/src/lib/renderer/pass-prop-decorator.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
export var passPropsSymbol = Symbol('passProps');
function passPropImpl(target, propertyKey, targetKey) {
    if (!target[passPropsSymbol]) {
        target[passPropsSymbol] = [];
    }
    target[passPropsSymbol].push({
        sourceKey: propertyKey,
        targetKey: targetKey,
    });
}
var passPropRename = function (propName) { return function (target, propertyKey) {
    return passPropImpl(target, propertyKey, propName);
}; };
var ɵ0 = passPropRename;
var passPropDirect = function (target, propertyKey) { return passPropImpl(target, propertyKey, propertyKey); };
var ɵ1 = passPropDirect;
export function passProp() {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    if (args.length === 0) {
        return passPropDirect;
    }
    if (args.length === 1) {
        return passPropRename(args[0]);
    }
    throw new Error('[passProp] Only 0 or 1 arguments accepted.');
}
export function getPassProps(obj) {
    return obj[passPropsSymbol];
}
export { ɵ0, ɵ1 };
