// @angular-react/core/esm5/src/lib/renderer/renderprop-helpers.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { createReactContentElement } from '../renderer/react-content';
import { createReactTemplateElement } from './react-template';
export function isRenderPropContext(x) {
    if (typeof x !== 'object') {
        return false;
    }
    var maybeRenderPropContext = x;
    return maybeRenderPropContext.render && typeof maybeRenderPropContext.render === 'function';
}
function renderReactContent(rootNodes, additionalProps) {
    return createReactContentElement(rootNodes, additionalProps);
}
/**
 * Wrap a `TemplateRef` with a `JSX.Element`.
 *
 * @param templateRef The template to wrap
 * @param ngZone A zone used for tracking & triggering updates to the template
 * @param additionalProps optional additional props to pass to the `ReactContent` object that will render the content.
 */
export function createTemplateRenderer(templateRef, ngZone, additionalProps) {
    return {
        render: function (context) { return createReactTemplateElement(templateRef, context, ngZone, additionalProps); },
    };
}
/**
 * Wrap a function resolving to an `HTMLElement` with a `JSX.Element`.
 *
 * @param htmlRenderFunc The function to wrap
 * @param additionalProps optional additional props to pass to the `ReactContent` object that will render the content.
 */
export function createHtmlRenderer(htmlRenderFunc, additionalProps) {
    return {
        render: function (context) {
            var rootHtmlElement = htmlRenderFunc(context);
            return renderReactContent([rootHtmlElement], additionalProps);
        },
    };
}
/**
 * Wrap a `ComponentRef` with a `JSX.Element`.
 *
 * @param htmlRenderFunc The component reference to wrap
 * @param additionalProps optional additional props to pass to the `ReactContent` object that will render the content.
 */
export function createComponentRenderer(componentRef, additionalProps) {
    var renderedJsx = null;
    return {
        render: function (context) {
            if (!renderedJsx) {
                renderedJsx = renderReactContent([componentRef.location.nativeElement], additionalProps);
            }
            Object.assign(componentRef.instance, context);
            componentRef.changeDetectorRef.detectChanges();
            return renderedJsx;
        },
    };
}
