// @wordpress/components/build/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  Animate: true,
  Autocomplete: true,
  BaseControl: true,
  Button: true,
  ButtonGroup: true,
  CheckboxControl: true,
  ClipboardButton: true,
  ColorIndicator: true,
  ColorPalette: true,
  ColorPicker: true,
  Dashicon: true,
  DateTimePicker: true,
  DatePicker: true,
  TimePicker: true,
  Disabled: true,
  Draggable: true,
  DropZone: true,
  DropZoneProvider: true,
  Dropdown: true,
  DropdownMenu: true,
  ExternalLink: true,
  FocalPointPicker: true,
  FocusableIframe: true,
  FontSizePicker: true,
  FormFileUpload: true,
  FormToggle: true,
  FormTokenField: true,
  Icon: true,
  IconButton: true,
  KeyboardShortcuts: true,
  MenuGroup: true,
  MenuItem: true,
  MenuItemsChoice: true,
  Modal: true,
  ScrollLock: true,
  NavigableMenu: true,
  TabbableContainer: true,
  Notice: true,
  NoticeList: true,
  Panel: true,
  PanelBody: true,
  PanelHeader: true,
  PanelRow: true,
  Placeholder: true,
  Popover: true,
  QueryControls: true,
  RadioControl: true,
  RangeControl: true,
  ResizableBox: true,
  ResponsiveWrapper: true,
  SandBox: true,
  SelectControl: true,
  Snackbar: true,
  SnackbarList: true,
  Spinner: true,
  TabPanel: true,
  TextControl: true,
  TextareaControl: true,
  Tip: true,
  ToggleControl: true,
  Toolbar: true,
  ToolbarButton: true,
  Tooltip: true,
  TreeSelect: true,
  IsolatedEventContainer: true,
  createSlotFill: true,
  Slot: true,
  Fill: true,
  SlotFillProvider: true,
  navigateRegions: true,
  withConstrainedTabbing: true,
  withFallbackStyles: true,
  withFilters: true,
  withFocusOutside: true,
  withFocusReturn: true,
  FocusReturnProvider: true,
  withNotices: true,
  withSpokenMessages: true
};
Object.defineProperty(exports, "Animate", {
  enumerable: true,
  get: function get() {
    return _animate.default;
  }
});
Object.defineProperty(exports, "Autocomplete", {
  enumerable: true,
  get: function get() {
    return _autocomplete.default;
  }
});
Object.defineProperty(exports, "BaseControl", {
  enumerable: true,
  get: function get() {
    return _baseControl.default;
  }
});
Object.defineProperty(exports, "Button", {
  enumerable: true,
  get: function get() {
    return _button.default;
  }
});
Object.defineProperty(exports, "ButtonGroup", {
  enumerable: true,
  get: function get() {
    return _buttonGroup.default;
  }
});
Object.defineProperty(exports, "CheckboxControl", {
  enumerable: true,
  get: function get() {
    return _checkboxControl.default;
  }
});
Object.defineProperty(exports, "ClipboardButton", {
  enumerable: true,
  get: function get() {
    return _clipboardButton.default;
  }
});
Object.defineProperty(exports, "ColorIndicator", {
  enumerable: true,
  get: function get() {
    return _colorIndicator.default;
  }
});
Object.defineProperty(exports, "ColorPalette", {
  enumerable: true,
  get: function get() {
    return _colorPalette.default;
  }
});
Object.defineProperty(exports, "ColorPicker", {
  enumerable: true,
  get: function get() {
    return _colorPicker.default;
  }
});
Object.defineProperty(exports, "Dashicon", {
  enumerable: true,
  get: function get() {
    return _dashicon.default;
  }
});
Object.defineProperty(exports, "DateTimePicker", {
  enumerable: true,
  get: function get() {
    return _dateTime.DateTimePicker;
  }
});
Object.defineProperty(exports, "DatePicker", {
  enumerable: true,
  get: function get() {
    return _dateTime.DatePicker;
  }
});
Object.defineProperty(exports, "TimePicker", {
  enumerable: true,
  get: function get() {
    return _dateTime.TimePicker;
  }
});
Object.defineProperty(exports, "Disabled", {
  enumerable: true,
  get: function get() {
    return _disabled.default;
  }
});
Object.defineProperty(exports, "Draggable", {
  enumerable: true,
  get: function get() {
    return _draggable.default;
  }
});
Object.defineProperty(exports, "DropZone", {
  enumerable: true,
  get: function get() {
    return _dropZone.default;
  }
});
Object.defineProperty(exports, "DropZoneProvider", {
  enumerable: true,
  get: function get() {
    return _provider.default;
  }
});
Object.defineProperty(exports, "Dropdown", {
  enumerable: true,
  get: function get() {
    return _dropdown.default;
  }
});
Object.defineProperty(exports, "DropdownMenu", {
  enumerable: true,
  get: function get() {
    return _dropdownMenu.default;
  }
});
Object.defineProperty(exports, "ExternalLink", {
  enumerable: true,
  get: function get() {
    return _externalLink.default;
  }
});
Object.defineProperty(exports, "FocalPointPicker", {
  enumerable: true,
  get: function get() {
    return _focalPointPicker.default;
  }
});
Object.defineProperty(exports, "FocusableIframe", {
  enumerable: true,
  get: function get() {
    return _focusableIframe.default;
  }
});
Object.defineProperty(exports, "FontSizePicker", {
  enumerable: true,
  get: function get() {
    return _fontSizePicker.default;
  }
});
Object.defineProperty(exports, "FormFileUpload", {
  enumerable: true,
  get: function get() {
    return _formFileUpload.default;
  }
});
Object.defineProperty(exports, "FormToggle", {
  enumerable: true,
  get: function get() {
    return _formToggle.default;
  }
});
Object.defineProperty(exports, "FormTokenField", {
  enumerable: true,
  get: function get() {
    return _formTokenField.default;
  }
});
Object.defineProperty(exports, "Icon", {
  enumerable: true,
  get: function get() {
    return _icon.default;
  }
});
Object.defineProperty(exports, "IconButton", {
  enumerable: true,
  get: function get() {
    return _iconButton.default;
  }
});
Object.defineProperty(exports, "KeyboardShortcuts", {
  enumerable: true,
  get: function get() {
    return _keyboardShortcuts.default;
  }
});
Object.defineProperty(exports, "MenuGroup", {
  enumerable: true,
  get: function get() {
    return _menuGroup.default;
  }
});
Object.defineProperty(exports, "MenuItem", {
  enumerable: true,
  get: function get() {
    return _menuItem.default;
  }
});
Object.defineProperty(exports, "MenuItemsChoice", {
  enumerable: true,
  get: function get() {
    return _menuItemsChoice.default;
  }
});
Object.defineProperty(exports, "Modal", {
  enumerable: true,
  get: function get() {
    return _modal.default;
  }
});
Object.defineProperty(exports, "ScrollLock", {
  enumerable: true,
  get: function get() {
    return _scrollLock.default;
  }
});
Object.defineProperty(exports, "NavigableMenu", {
  enumerable: true,
  get: function get() {
    return _navigableContainer.NavigableMenu;
  }
});
Object.defineProperty(exports, "TabbableContainer", {
  enumerable: true,
  get: function get() {
    return _navigableContainer.TabbableContainer;
  }
});
Object.defineProperty(exports, "Notice", {
  enumerable: true,
  get: function get() {
    return _notice.default;
  }
});
Object.defineProperty(exports, "NoticeList", {
  enumerable: true,
  get: function get() {
    return _list.default;
  }
});
Object.defineProperty(exports, "Panel", {
  enumerable: true,
  get: function get() {
    return _panel.default;
  }
});
Object.defineProperty(exports, "PanelBody", {
  enumerable: true,
  get: function get() {
    return _body.default;
  }
});
Object.defineProperty(exports, "PanelHeader", {
  enumerable: true,
  get: function get() {
    return _header.default;
  }
});
Object.defineProperty(exports, "PanelRow", {
  enumerable: true,
  get: function get() {
    return _row.default;
  }
});
Object.defineProperty(exports, "Placeholder", {
  enumerable: true,
  get: function get() {
    return _placeholder.default;
  }
});
Object.defineProperty(exports, "Popover", {
  enumerable: true,
  get: function get() {
    return _popover.default;
  }
});
Object.defineProperty(exports, "QueryControls", {
  enumerable: true,
  get: function get() {
    return _queryControls.default;
  }
});
Object.defineProperty(exports, "RadioControl", {
  enumerable: true,
  get: function get() {
    return _radioControl.default;
  }
});
Object.defineProperty(exports, "RangeControl", {
  enumerable: true,
  get: function get() {
    return _rangeControl.default;
  }
});
Object.defineProperty(exports, "ResizableBox", {
  enumerable: true,
  get: function get() {
    return _resizableBox.default;
  }
});
Object.defineProperty(exports, "ResponsiveWrapper", {
  enumerable: true,
  get: function get() {
    return _responsiveWrapper.default;
  }
});
Object.defineProperty(exports, "SandBox", {
  enumerable: true,
  get: function get() {
    return _sandbox.default;
  }
});
Object.defineProperty(exports, "SelectControl", {
  enumerable: true,
  get: function get() {
    return _selectControl.default;
  }
});
Object.defineProperty(exports, "Snackbar", {
  enumerable: true,
  get: function get() {
    return _snackbar.default;
  }
});
Object.defineProperty(exports, "SnackbarList", {
  enumerable: true,
  get: function get() {
    return _list2.default;
  }
});
Object.defineProperty(exports, "Spinner", {
  enumerable: true,
  get: function get() {
    return _spinner.default;
  }
});
Object.defineProperty(exports, "TabPanel", {
  enumerable: true,
  get: function get() {
    return _tabPanel.default;
  }
});
Object.defineProperty(exports, "TextControl", {
  enumerable: true,
  get: function get() {
    return _textControl.default;
  }
});
Object.defineProperty(exports, "TextareaControl", {
  enumerable: true,
  get: function get() {
    return _textareaControl.default;
  }
});
Object.defineProperty(exports, "Tip", {
  enumerable: true,
  get: function get() {
    return _tip.default;
  }
});
Object.defineProperty(exports, "ToggleControl", {
  enumerable: true,
  get: function get() {
    return _toggleControl.default;
  }
});
Object.defineProperty(exports, "Toolbar", {
  enumerable: true,
  get: function get() {
    return _toolbar.default;
  }
});
Object.defineProperty(exports, "ToolbarButton", {
  enumerable: true,
  get: function get() {
    return _toolbarButton.default;
  }
});
Object.defineProperty(exports, "Tooltip", {
  enumerable: true,
  get: function get() {
    return _tooltip.default;
  }
});
Object.defineProperty(exports, "TreeSelect", {
  enumerable: true,
  get: function get() {
    return _treeSelect.default;
  }
});
Object.defineProperty(exports, "IsolatedEventContainer", {
  enumerable: true,
  get: function get() {
    return _isolatedEventContainer.default;
  }
});
Object.defineProperty(exports, "createSlotFill", {
  enumerable: true,
  get: function get() {
    return _slotFill.createSlotFill;
  }
});
Object.defineProperty(exports, "Slot", {
  enumerable: true,
  get: function get() {
    return _slotFill.Slot;
  }
});
Object.defineProperty(exports, "Fill", {
  enumerable: true,
  get: function get() {
    return _slotFill.Fill;
  }
});
Object.defineProperty(exports, "SlotFillProvider", {
  enumerable: true,
  get: function get() {
    return _slotFill.Provider;
  }
});
Object.defineProperty(exports, "navigateRegions", {
  enumerable: true,
  get: function get() {
    return _navigateRegions.default;
  }
});
Object.defineProperty(exports, "withConstrainedTabbing", {
  enumerable: true,
  get: function get() {
    return _withConstrainedTabbing.default;
  }
});
Object.defineProperty(exports, "withFallbackStyles", {
  enumerable: true,
  get: function get() {
    return _withFallbackStyles.default;
  }
});
Object.defineProperty(exports, "withFilters", {
  enumerable: true,
  get: function get() {
    return _withFilters.default;
  }
});
Object.defineProperty(exports, "withFocusOutside", {
  enumerable: true,
  get: function get() {
    return _withFocusOutside.default;
  }
});
Object.defineProperty(exports, "withFocusReturn", {
  enumerable: true,
  get: function get() {
    return _withFocusReturn.default;
  }
});
Object.defineProperty(exports, "FocusReturnProvider", {
  enumerable: true,
  get: function get() {
    return _withFocusReturn.Provider;
  }
});
Object.defineProperty(exports, "withNotices", {
  enumerable: true,
  get: function get() {
    return _withNotices.default;
  }
});
Object.defineProperty(exports, "withSpokenMessages", {
  enumerable: true,
  get: function get() {
    return _withSpokenMessages.default;
  }
});

var _primitives = require("./primitives");

Object.keys(_primitives).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _primitives[key];
    }
  });
});

var _animate = _interopRequireDefault(require("./animate"));

var _autocomplete = _interopRequireDefault(require("./autocomplete"));

var _baseControl = _interopRequireDefault(require("./base-control"));

var _button = _interopRequireDefault(require("./button"));

var _buttonGroup = _interopRequireDefault(require("./button-group"));

var _checkboxControl = _interopRequireDefault(require("./checkbox-control"));

var _clipboardButton = _interopRequireDefault(require("./clipboard-button"));

var _colorIndicator = _interopRequireDefault(require("./color-indicator"));

var _colorPalette = _interopRequireDefault(require("./color-palette"));

var _colorPicker = _interopRequireDefault(require("./color-picker"));

var _dashicon = _interopRequireDefault(require("./dashicon"));

var _dateTime = require("./date-time");

var _disabled = _interopRequireDefault(require("./disabled"));

var _draggable = _interopRequireDefault(require("./draggable"));

var _dropZone = _interopRequireDefault(require("./drop-zone"));

var _provider = _interopRequireDefault(require("./drop-zone/provider"));

var _dropdown = _interopRequireDefault(require("./dropdown"));

var _dropdownMenu = _interopRequireDefault(require("./dropdown-menu"));

var _externalLink = _interopRequireDefault(require("./external-link"));

var _focalPointPicker = _interopRequireDefault(require("./focal-point-picker"));

var _focusableIframe = _interopRequireDefault(require("./focusable-iframe"));

var _fontSizePicker = _interopRequireDefault(require("./font-size-picker"));

var _formFileUpload = _interopRequireDefault(require("./form-file-upload"));

var _formToggle = _interopRequireDefault(require("./form-toggle"));

var _formTokenField = _interopRequireDefault(require("./form-token-field"));

var _icon = _interopRequireDefault(require("./icon"));

var _iconButton = _interopRequireDefault(require("./icon-button"));

var _keyboardShortcuts = _interopRequireDefault(require("./keyboard-shortcuts"));

var _menuGroup = _interopRequireDefault(require("./menu-group"));

var _menuItem = _interopRequireDefault(require("./menu-item"));

var _menuItemsChoice = _interopRequireDefault(require("./menu-items-choice"));

var _modal = _interopRequireDefault(require("./modal"));

var _scrollLock = _interopRequireDefault(require("./scroll-lock"));

var _navigableContainer = require("./navigable-container");

var _notice = _interopRequireDefault(require("./notice"));

var _list = _interopRequireDefault(require("./notice/list"));

var _panel = _interopRequireDefault(require("./panel"));

var _body = _interopRequireDefault(require("./panel/body"));

var _header = _interopRequireDefault(require("./panel/header"));

var _row = _interopRequireDefault(require("./panel/row"));

var _placeholder = _interopRequireDefault(require("./placeholder"));

var _popover = _interopRequireDefault(require("./popover"));

var _queryControls = _interopRequireDefault(require("./query-controls"));

var _radioControl = _interopRequireDefault(require("./radio-control"));

var _rangeControl = _interopRequireDefault(require("./range-control"));

var _resizableBox = _interopRequireDefault(require("./resizable-box"));

var _responsiveWrapper = _interopRequireDefault(require("./responsive-wrapper"));

var _sandbox = _interopRequireDefault(require("./sandbox"));

var _selectControl = _interopRequireDefault(require("./select-control"));

var _snackbar = _interopRequireDefault(require("./snackbar"));

var _list2 = _interopRequireDefault(require("./snackbar/list"));

var _spinner = _interopRequireDefault(require("./spinner"));

var _tabPanel = _interopRequireDefault(require("./tab-panel"));

var _textControl = _interopRequireDefault(require("./text-control"));

var _textareaControl = _interopRequireDefault(require("./textarea-control"));

var _tip = _interopRequireDefault(require("./tip"));

var _toggleControl = _interopRequireDefault(require("./toggle-control"));

var _toolbar = _interopRequireDefault(require("./toolbar"));

var _toolbarButton = _interopRequireDefault(require("./toolbar-button"));

var _tooltip = _interopRequireDefault(require("./tooltip"));

var _treeSelect = _interopRequireDefault(require("./tree-select"));

var _isolatedEventContainer = _interopRequireDefault(require("./isolated-event-container"));

var _slotFill = require("./slot-fill");

var _navigateRegions = _interopRequireDefault(require("./higher-order/navigate-regions"));

var _withConstrainedTabbing = _interopRequireDefault(require("./higher-order/with-constrained-tabbing"));

var _withFallbackStyles = _interopRequireDefault(require("./higher-order/with-fallback-styles"));

var _withFilters = _interopRequireDefault(require("./higher-order/with-filters"));

var _withFocusOutside = _interopRequireDefault(require("./higher-order/with-focus-outside"));

var _withFocusReturn = _interopRequireWildcard(require("./higher-order/with-focus-return"));

var _withNotices = _interopRequireDefault(require("./higher-order/with-notices"));

var _withSpokenMessages = _interopRequireDefault(require("./higher-order/with-spoken-messages"));
