// @wordpress/components/build/font-size-picker/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _i18n = require("@wordpress/i18n");

var _button = _interopRequireDefault(require("../button"));

var _rangeControl = _interopRequireDefault(require("../range-control"));

var _selectControl = _interopRequireDefault(require("../select-control"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function getSelectValueFromFontSize(fontSizes, value) {
  if (value) {
    var fontSizeValue = fontSizes.find(function (font) {
      return font.size === value;
    });
    return fontSizeValue ? fontSizeValue.slug : 'custom';
  }

  return 'normal';
}

function getSelectOptions(optionsArray) {
  return [].concat((0, _toConsumableArray2.default)(optionsArray.map(function (option) {
    return {
      value: option.slug,
      label: option.name
    };
  })), [{
    value: 'custom',
    label: (0, _i18n.__)('Custom')
  }]);
}

function FontSizePicker(_ref) {
  var fallbackFontSize = _ref.fallbackFontSize,
      _ref$fontSizes = _ref.fontSizes,
      fontSizes = _ref$fontSizes === void 0 ? [] : _ref$fontSizes,
      _ref$disableCustomFon = _ref.disableCustomFontSizes,
      disableCustomFontSizes = _ref$disableCustomFon === void 0 ? false : _ref$disableCustomFon,
      onChange = _ref.onChange,
      value = _ref.value,
      _ref$withSlider = _ref.withSlider,
      withSlider = _ref$withSlider === void 0 ? false : _ref$withSlider;

  var _useState = (0, _element.useState)(getSelectValueFromFontSize(fontSizes, value)),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      currentSelectValue = _useState2[0],
      setCurrentSelectValue = _useState2[1];

  if (disableCustomFontSizes && !fontSizes.length) {
    return null;
  }

  var onChangeValue = function onChangeValue(event) {
    var newValue = event.target.value;
    setCurrentSelectValue(getSelectValueFromFontSize(fontSizes, Number(newValue)));

    if (newValue === '') {
      onChange(undefined);
      return;
    }

    onChange(Number(newValue));
  };

  var onSelectChangeValue = function onSelectChangeValue(eventValue) {
    setCurrentSelectValue(eventValue);
    var selectedFont = fontSizes.find(function (font) {
      return font.slug === eventValue;
    });

    if (selectedFont) {
      onChange(selectedFont.size);
    }
  };

  return (0, _element.createElement)("fieldset", null, (0, _element.createElement)("legend", null, (0, _i18n.__)('Font Size')), (0, _element.createElement)("div", {
    className: "components-font-size-picker__controls"
  }, fontSizes.length > 0 && (0, _element.createElement)(_selectControl.default, {
    className: 'components-font-size-picker__select',
    label: 'Choose preset',
    hideLabelFromVision: true,
    value: currentSelectValue,
    onChange: onSelectChangeValue,
    options: getSelectOptions(fontSizes)
  }), !withSlider && !disableCustomFontSizes && (0, _element.createElement)("input", {
    className: "components-range-control__number",
    type: "number",
    onChange: onChangeValue,
    "aria-label": (0, _i18n.__)('Custom'),
    value: value || ''
  }), (0, _element.createElement)(_button.default, {
    className: "components-color-palette__clear",
    type: "button",
    disabled: value === undefined,
    onClick: function onClick() {
      onChange(undefined);
      setCurrentSelectValue(getSelectValueFromFontSize(fontSizes, undefined));
    },
    isSmall: true,
    isDefault: true
  }, (0, _i18n.__)('Reset'))), withSlider && (0, _element.createElement)(_rangeControl.default, {
    className: "components-font-size-picker__custom-input",
    label: (0, _i18n.__)('Custom Size'),
    value: value || '',
    initialPosition: fallbackFontSize,
    onChange: onChange,
    min: 12,
    max: 100,
    beforeIcon: "editor-textcolor",
    afterIcon: "editor-textcolor"
  }));
}

var _default = FontSizePicker;
exports.default = _default;
