// @wordpress/editor/build-module/store/reducer.native.js
/**
 * External dependencies
 */
import optimist from 'redux-optimist';
/**
 * WordPress dependencies
 */

import { combineReducers } from '@wordpress/data';
/**
 * Internal dependencies
 */

import { editor, initialEdits, currentPost, preferences, saving, postLock, reusableBlocks, template, previewLink, postSavingLock, isReady, editorSettings } from './reducer.js';
export * from './reducer.js';
/**
 * Reducer returning the post title state.
 *
 * @param {PostTitleState} state  Current state.
 * @param {Object}         action Dispatched action.
 *
 * @return {Object} Updated state.
 */

export var postTitle = combineReducers({
  isSelected: function isSelected() {
    var state = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;
    var action = arguments.length > 1 ? arguments[1] : undefined;

    switch (action.type) {
      case 'TOGGLE_POST_TITLE_SELECTION':
        return action.isSelected;
    }

    return state;
  }
});
export default optimist(combineReducers({
  editor: editor,
  initialEdits: initialEdits,
  currentPost: currentPost,
  preferences: preferences,
  saving: saving,
  postLock: postLock,
  reusableBlocks: reusableBlocks,
  template: template,
  previewLink: previewLink,
  postSavingLock: postSavingLock,
  isReady: isReady,
  editorSettings: editorSettings,
  postTitle: postTitle
}));
