// @wordpress/block-editor/build/components/url-popover/link-viewer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = LinkViewer;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classnames = _interopRequireDefault(require("classnames"));

var _i18n = require("@wordpress/i18n");

var _components = require("@wordpress/components");

var _url = require("@wordpress/url");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */
function LinkViewerUrl(_ref) {
  var url = _ref.url,
      urlLabel = _ref.urlLabel,
      className = _ref.className;
  var linkClassName = (0, _classnames.default)(className, 'block-editor-url-popover__link-viewer-url');

  if (!url) {
    return (0, _element.createElement)("span", {
      className: linkClassName
    });
  }

  return (0, _element.createElement)(_components.ExternalLink, {
    className: linkClassName,
    href: url
  }, urlLabel || (0, _url.filterURLForDisplay)((0, _url.safeDecodeURI)(url)));
}

function LinkViewer(_ref2) {
  var className = _ref2.className,
      linkClassName = _ref2.linkClassName,
      onEditLinkClick = _ref2.onEditLinkClick,
      url = _ref2.url,
      urlLabel = _ref2.urlLabel,
      props = (0, _objectWithoutProperties2.default)(_ref2, ["className", "linkClassName", "onEditLinkClick", "url", "urlLabel"]);
  return (0, _element.createElement)("div", (0, _extends2.default)({
    className: (0, _classnames.default)('block-editor-url-popover__link-viewer', className)
  }, props), (0, _element.createElement)(LinkViewerUrl, {
    url: url,
    urlLabel: urlLabel,
    className: linkClassName
  }), onEditLinkClick && (0, _element.createElement)(_components.IconButton, {
    icon: "edit",
    label: (0, _i18n.__)('Edit'),
    onClick: onEditLinkClick
  }));
}
