// @wordpress/block-library/build-module/paragraph/edit.native.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _assertThisInitialized from "@babel/runtime/helpers/esm/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { View } from 'react-native';
/**
 * WordPress dependencies
 */

import { __ } from '@wordpress/i18n';
import { Component } from '@wordpress/element';
import { createBlock } from '@wordpress/blocks';
import { RichText } from '@wordpress/block-editor';
/**
 * Internal dependencies
 */

var name = 'core/paragraph';

var ParagraphEdit =
/*#__PURE__*/
function (_Component) {
  _inherits(ParagraphEdit, _Component);

  function ParagraphEdit(props) {
    var _this;

    _classCallCheck(this, ParagraphEdit);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(ParagraphEdit).call(this, props));
    _this.onReplace = _this.onReplace.bind(_assertThisInitialized(_this));
    return _this;
  }

  _createClass(ParagraphEdit, [{
    key: "onReplace",
    value: function onReplace(blocks) {
      var _this$props = this.props,
          attributes = _this$props.attributes,
          onReplace = _this$props.onReplace;
      onReplace(blocks.map(function (block, index) {
        return index === 0 && block.name === name ? _objectSpread({}, block, {
          attributes: _objectSpread({}, attributes, block.attributes)
        }) : block;
      }));
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          attributes = _this$props2.attributes,
          setAttributes = _this$props2.setAttributes,
          mergeBlocks = _this$props2.mergeBlocks,
          onReplace = _this$props2.onReplace,
          style = _this$props2.style;
      var placeholder = attributes.placeholder,
          content = attributes.content;
      return createElement(View, null, createElement(RichText, {
        identifier: "content",
        tagName: "p",
        value: content,
        deleteEnter: true,
        style: style,
        onChange: function onChange(nextContent) {
          setAttributes({
            content: nextContent
          });
        },
        onSplit: function onSplit(value) {
          if (!value) {
            return createBlock(name);
          }

          return createBlock(name, _objectSpread({}, attributes, {
            content: value
          }));
        },
        onMerge: mergeBlocks,
        onReplace: onReplace,
        onRemove: onReplace ? function () {
          return onReplace([]);
        } : undefined,
        placeholder: placeholder || __('Start writing…')
      }));
    }
  }]);

  return ParagraphEdit;
}(Component);

export default ParagraphEdit;
