// @wordpress/components/build/drop-zone/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _i18n = require("@wordpress/i18n");

var _dashicon = _interopRequireDefault(require("../dashicon"));

var _provider = require("./provider");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var DropZone = function DropZone(props) {
  return (0, _element.createElement)(_provider.DropZoneConsumer, null, function (_ref) {
    var addDropZone = _ref.addDropZone,
        removeDropZone = _ref.removeDropZone;
    return (0, _element.createElement)(DropZoneComponent, (0, _extends2.default)({
      addDropZone: addDropZone,
      removeDropZone: removeDropZone
    }, props));
  });
};

var DropZoneComponent =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(DropZoneComponent, _Component);

  function DropZoneComponent() {
    var _this;

    (0, _classCallCheck2.default)(this, DropZoneComponent);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(DropZoneComponent).apply(this, arguments));
    _this.dropZoneElement = (0, _element.createRef)();
    _this.dropZone = {
      element: null,
      onDrop: _this.props.onDrop,
      onFilesDrop: _this.props.onFilesDrop,
      onHTMLDrop: _this.props.onHTMLDrop,
      setState: _this.setState.bind((0, _assertThisInitialized2.default)(_this))
    };
    _this.state = {
      isDraggingOverDocument: false,
      isDraggingOverElement: false,
      position: null,
      type: null
    };
    return _this;
  }

  (0, _createClass2.default)(DropZoneComponent, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      // Set element after the component has a node assigned in the DOM
      this.dropZone.element = this.dropZoneElement.current;
      this.props.addDropZone(this.dropZone);
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.props.removeDropZone(this.dropZone);
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          className = _this$props.className,
          label = _this$props.label;
      var _this$state = this.state,
          isDraggingOverDocument = _this$state.isDraggingOverDocument,
          isDraggingOverElement = _this$state.isDraggingOverElement,
          position = _this$state.position,
          type = _this$state.type;
      var classes = (0, _classnames2.default)('components-drop-zone', className, (0, _defineProperty2.default)({
        'is-active': isDraggingOverDocument || isDraggingOverElement,
        'is-dragging-over-document': isDraggingOverDocument,
        'is-dragging-over-element': isDraggingOverElement,
        'is-close-to-top': position && position.y === 'top',
        'is-close-to-bottom': position && position.y === 'bottom',
        'is-close-to-left': position && position.x === 'left',
        'is-close-to-right': position && position.x === 'right'
      }, "is-dragging-".concat(type), !!type));
      var children;

      if (isDraggingOverElement) {
        children = (0, _element.createElement)("div", {
          className: "components-drop-zone__content"
        }, (0, _element.createElement)(_dashicon.default, {
          icon: "upload",
          size: "40",
          className: "components-drop-zone__content-icon"
        }), (0, _element.createElement)("span", {
          className: "components-drop-zone__content-text"
        }, label ? label : (0, _i18n.__)('Drop files to upload')));
      }

      return (0, _element.createElement)("div", {
        ref: this.dropZoneElement,
        className: classes
      }, children);
    }
  }]);
  return DropZoneComponent;
}(_element.Component);

var _default = DropZone;
exports.default = _default;
