// @wordpress/editor/build/components/deprecated.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ServerSideRender", {
  enumerable: true,
  get: function get() {
    return _serverSideRender.default;
  }
});
exports.withFontSizes = exports.withColors = exports.withColorContext = exports.getFontSizeClass = exports.getFontSize = exports.getColorObjectByColorValue = exports.getColorObjectByAttributeValues = exports.getColorClassName = exports.createCustomColorsHOC = exports.WritingFlow = exports.Warning = exports.URLPopover = exports.URLInputButton = exports.URLInput = exports.SkipToSelectedBlock = exports.PreserveScrollInReorder = exports.ObserveTyping = exports.NavigableToolbar = exports.MultiSelectScrollIntoView = exports.MultiBlocksSwitcher = exports.MediaUploadCheck = exports.MediaUpload = exports.MediaPlaceholder = exports.__unstableRichTextInputEvent = exports.RichTextToolbarButton = exports.RichTextShortcut = exports.PlainText = exports.PanelColorSettings = exports.InspectorControls = exports.InspectorAdvancedControls = exports.InnerBlocks = exports.Inserter = exports.FontSizePicker = exports.DefaultBlockAppender = exports.CopyHandler = exports.ContrastChecker = exports.ColorPalette = exports.BlockToolbar = exports.BlockTitle = exports.BlockSettingsMenu = exports.BlockSelectionClearer = exports.BlockNavigationDropdown = exports.BlockMover = exports.BlockList = exports.BlockInspector = exports.BlockIcon = exports.BlockFormatControls = exports.BlockEditorKeyboardShortcuts = exports.BlockEdit = exports.BlockControls = exports.BlockAlignmentToolbar = exports.AlignmentToolbar = exports.Autocomplete = exports.RichText = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _deprecated = _interopRequireDefault(require("@wordpress/deprecated"));

var _blockEditor = require("@wordpress/block-editor");

var _serverSideRender = _interopRequireDefault(require("@wordpress/server-side-render"));

// Block Creation Components

/**
 * WordPress dependencies
 */
function deprecateComponent(name, Wrapped) {
  var staticsToHoist = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : [];
  var Component = (0, _element.forwardRef)(function (props, ref) {
    (0, _deprecated.default)('wp.editor.' + name, {
      alternative: 'wp.blockEditor.' + name
    });
    return (0, _element.createElement)(Wrapped, (0, _extends2.default)({
      ref: ref
    }, props));
  });
  staticsToHoist.forEach(function (staticName) {
    Component[staticName] = deprecateComponent(name + '.' + staticName, Wrapped[staticName]);
  });
  return Component;
}

function deprecateFunction(name, func) {
  return function () {
    (0, _deprecated.default)('wp.editor.' + name, {
      alternative: 'wp.blockEditor.' + name
    });
    return func.apply(void 0, arguments);
  };
}

var RichText = deprecateComponent('RichText', _blockEditor.RichText, ['Content']);
exports.RichText = RichText;
RichText.isEmpty = deprecateFunction('RichText.isEmpty', _blockEditor.RichText.isEmpty);
var Autocomplete = deprecateComponent('Autocomplete', _blockEditor.Autocomplete);
exports.Autocomplete = Autocomplete;
var AlignmentToolbar = deprecateComponent('AlignmentToolbar', _blockEditor.AlignmentToolbar);
exports.AlignmentToolbar = AlignmentToolbar;
var BlockAlignmentToolbar = deprecateComponent('BlockAlignmentToolbar', _blockEditor.BlockAlignmentToolbar);
exports.BlockAlignmentToolbar = BlockAlignmentToolbar;
var BlockControls = deprecateComponent('BlockControls', _blockEditor.BlockControls, ['Slot']);
exports.BlockControls = BlockControls;
var BlockEdit = deprecateComponent('BlockEdit', _blockEditor.BlockEdit);
exports.BlockEdit = BlockEdit;
var BlockEditorKeyboardShortcuts = deprecateComponent('BlockEditorKeyboardShortcuts', _blockEditor.BlockEditorKeyboardShortcuts);
exports.BlockEditorKeyboardShortcuts = BlockEditorKeyboardShortcuts;
var BlockFormatControls = deprecateComponent('BlockFormatControls', _blockEditor.BlockFormatControls, ['Slot']);
exports.BlockFormatControls = BlockFormatControls;
var BlockIcon = deprecateComponent('BlockIcon', _blockEditor.BlockIcon);
exports.BlockIcon = BlockIcon;
var BlockInspector = deprecateComponent('BlockInspector', _blockEditor.BlockInspector);
exports.BlockInspector = BlockInspector;
var BlockList = deprecateComponent('BlockList', _blockEditor.BlockList);
exports.BlockList = BlockList;
var BlockMover = deprecateComponent('BlockMover', _blockEditor.BlockMover);
exports.BlockMover = BlockMover;
var BlockNavigationDropdown = deprecateComponent('BlockNavigationDropdown', _blockEditor.BlockNavigationDropdown);
exports.BlockNavigationDropdown = BlockNavigationDropdown;
var BlockSelectionClearer = deprecateComponent('BlockSelectionClearer', _blockEditor.BlockSelectionClearer);
exports.BlockSelectionClearer = BlockSelectionClearer;
var BlockSettingsMenu = deprecateComponent('BlockSettingsMenu', _blockEditor.BlockSettingsMenu);
exports.BlockSettingsMenu = BlockSettingsMenu;
var BlockTitle = deprecateComponent('BlockTitle', _blockEditor.BlockTitle);
exports.BlockTitle = BlockTitle;
var BlockToolbar = deprecateComponent('BlockToolbar', _blockEditor.BlockToolbar);
exports.BlockToolbar = BlockToolbar;
var ColorPalette = deprecateComponent('ColorPalette', _blockEditor.ColorPalette);
exports.ColorPalette = ColorPalette;
var ContrastChecker = deprecateComponent('ContrastChecker', _blockEditor.ContrastChecker);
exports.ContrastChecker = ContrastChecker;
var CopyHandler = deprecateComponent('CopyHandler', _blockEditor.CopyHandler);
exports.CopyHandler = CopyHandler;
var DefaultBlockAppender = deprecateComponent('DefaultBlockAppender', _blockEditor.DefaultBlockAppender);
exports.DefaultBlockAppender = DefaultBlockAppender;
var FontSizePicker = deprecateComponent('FontSizePicker', _blockEditor.FontSizePicker);
exports.FontSizePicker = FontSizePicker;
var Inserter = deprecateComponent('Inserter', _blockEditor.Inserter);
exports.Inserter = Inserter;
var InnerBlocks = deprecateComponent('InnerBlocks', _blockEditor.InnerBlocks, ['ButtonBlockAppender', 'DefaultBlockAppender', 'Content']);
exports.InnerBlocks = InnerBlocks;
var InspectorAdvancedControls = deprecateComponent('InspectorAdvancedControls', _blockEditor.InspectorAdvancedControls, ['Slot']);
exports.InspectorAdvancedControls = InspectorAdvancedControls;
var InspectorControls = deprecateComponent('InspectorControls', _blockEditor.InspectorControls, ['Slot']);
exports.InspectorControls = InspectorControls;
var PanelColorSettings = deprecateComponent('PanelColorSettings', _blockEditor.PanelColorSettings);
exports.PanelColorSettings = PanelColorSettings;
var PlainText = deprecateComponent('PlainText', _blockEditor.PlainText);
exports.PlainText = PlainText;
var RichTextShortcut = deprecateComponent('RichTextShortcut', _blockEditor.RichTextShortcut);
exports.RichTextShortcut = RichTextShortcut;
var RichTextToolbarButton = deprecateComponent('RichTextToolbarButton', _blockEditor.RichTextToolbarButton);
exports.RichTextToolbarButton = RichTextToolbarButton;

var __unstableRichTextInputEvent = deprecateComponent('__unstableRichTextInputEvent', _blockEditor.__unstableRichTextInputEvent);

exports.__unstableRichTextInputEvent = __unstableRichTextInputEvent;
var MediaPlaceholder = deprecateComponent('MediaPlaceholder', _blockEditor.MediaPlaceholder);
exports.MediaPlaceholder = MediaPlaceholder;
var MediaUpload = deprecateComponent('MediaUpload', _blockEditor.MediaUpload);
exports.MediaUpload = MediaUpload;
var MediaUploadCheck = deprecateComponent('MediaUploadCheck', _blockEditor.MediaUploadCheck);
exports.MediaUploadCheck = MediaUploadCheck;
var MultiBlocksSwitcher = deprecateComponent('MultiBlocksSwitcher', _blockEditor.MultiBlocksSwitcher);
exports.MultiBlocksSwitcher = MultiBlocksSwitcher;
var MultiSelectScrollIntoView = deprecateComponent('MultiSelectScrollIntoView', _blockEditor.MultiSelectScrollIntoView);
exports.MultiSelectScrollIntoView = MultiSelectScrollIntoView;
var NavigableToolbar = deprecateComponent('NavigableToolbar', _blockEditor.NavigableToolbar);
exports.NavigableToolbar = NavigableToolbar;
var ObserveTyping = deprecateComponent('ObserveTyping', _blockEditor.ObserveTyping);
exports.ObserveTyping = ObserveTyping;
var PreserveScrollInReorder = deprecateComponent('PreserveScrollInReorder', _blockEditor.PreserveScrollInReorder);
exports.PreserveScrollInReorder = PreserveScrollInReorder;
var SkipToSelectedBlock = deprecateComponent('SkipToSelectedBlock', _blockEditor.SkipToSelectedBlock);
exports.SkipToSelectedBlock = SkipToSelectedBlock;
var URLInput = deprecateComponent('URLInput', _blockEditor.URLInput);
exports.URLInput = URLInput;
var URLInputButton = deprecateComponent('URLInputButton', _blockEditor.URLInputButton);
exports.URLInputButton = URLInputButton;
var URLPopover = deprecateComponent('URLPopover', _blockEditor.URLPopover);
exports.URLPopover = URLPopover;
var Warning = deprecateComponent('Warning', _blockEditor.Warning);
exports.Warning = Warning;
var WritingFlow = deprecateComponent('WritingFlow', _blockEditor.WritingFlow);
exports.WritingFlow = WritingFlow;
var createCustomColorsHOC = deprecateFunction('createCustomColorsHOC', _blockEditor.createCustomColorsHOC);
exports.createCustomColorsHOC = createCustomColorsHOC;
var getColorClassName = deprecateFunction('getColorClassName', _blockEditor.getColorClassName);
exports.getColorClassName = getColorClassName;
var getColorObjectByAttributeValues = deprecateFunction('getColorObjectByAttributeValues', _blockEditor.getColorObjectByAttributeValues);
exports.getColorObjectByAttributeValues = getColorObjectByAttributeValues;
var getColorObjectByColorValue = deprecateFunction('getColorObjectByColorValue', _blockEditor.getColorObjectByColorValue);
exports.getColorObjectByColorValue = getColorObjectByColorValue;
var getFontSize = deprecateFunction('getFontSize', _blockEditor.getFontSize);
exports.getFontSize = getFontSize;
var getFontSizeClass = deprecateFunction('getFontSizeClass', _blockEditor.getFontSizeClass);
exports.getFontSizeClass = getFontSizeClass;
var withColorContext = deprecateFunction('withColorContext', _blockEditor.withColorContext);
exports.withColorContext = withColorContext;
var withColors = deprecateFunction('withColors', _blockEditor.withColors);
exports.withColors = withColors;
var withFontSizes = deprecateFunction('withFontSizes', _blockEditor.withFontSizes);
exports.withFontSizes = withFontSizes;
