// @wordpress/editor/build/components/deprecated.native.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "BlockControls", {
  enumerable: true,
  get: function get() {
    return _blockEditor.BlockControls;
  }
});
Object.defineProperty(exports, "BlockEdit", {
  enumerable: true,
  get: function get() {
    return _blockEditor.BlockEdit;
  }
});
Object.defineProperty(exports, "BlockFormatControls", {
  enumerable: true,
  get: function get() {
    return _blockEditor.BlockFormatControls;
  }
});
Object.defineProperty(exports, "DefaultBlockAppender", {
  enumerable: true,
  get: function get() {
    return _blockEditor.DefaultBlockAppender;
  }
});
Object.defineProperty(exports, "createCustomColorsHOC", {
  enumerable: true,
  get: function get() {
    return _blockEditor.createCustomColorsHOC;
  }
});
Object.defineProperty(exports, "getColorClassName", {
  enumerable: true,
  get: function get() {
    return _blockEditor.getColorClassName;
  }
});
Object.defineProperty(exports, "getColorObjectByAttributeValues", {
  enumerable: true,
  get: function get() {
    return _blockEditor.getColorObjectByAttributeValues;
  }
});
Object.defineProperty(exports, "getColorObjectByColorValue", {
  enumerable: true,
  get: function get() {
    return _blockEditor.getColorObjectByColorValue;
  }
});
Object.defineProperty(exports, "InspectorControls", {
  enumerable: true,
  get: function get() {
    return _blockEditor.InspectorControls;
  }
});
Object.defineProperty(exports, "PlainText", {
  enumerable: true,
  get: function get() {
    return _blockEditor.PlainText;
  }
});
Object.defineProperty(exports, "RichText", {
  enumerable: true,
  get: function get() {
    return _blockEditor.RichText;
  }
});
Object.defineProperty(exports, "RichTextShortcut", {
  enumerable: true,
  get: function get() {
    return _blockEditor.RichTextShortcut;
  }
});
Object.defineProperty(exports, "RichTextToolbarButton", {
  enumerable: true,
  get: function get() {
    return _blockEditor.RichTextToolbarButton;
  }
});
Object.defineProperty(exports, "RichTextInserterItem", {
  enumerable: true,
  get: function get() {
    return _blockEditor.RichTextInserterItem;
  }
});
Object.defineProperty(exports, "__unstableRichTextInputEvent", {
  enumerable: true,
  get: function get() {
    return _blockEditor.__unstableRichTextInputEvent;
  }
});
Object.defineProperty(exports, "MediaPlaceholder", {
  enumerable: true,
  get: function get() {
    return _blockEditor.MediaPlaceholder;
  }
});
Object.defineProperty(exports, "URLInput", {
  enumerable: true,
  get: function get() {
    return _blockEditor.URLInput;
  }
});
Object.defineProperty(exports, "withColors", {
  enumerable: true,
  get: function get() {
    return _blockEditor.withColors;
  }
});
Object.defineProperty(exports, "withFontSizes", {
  enumerable: true,
  get: function get() {
    return _blockEditor.withFontSizes;
  }
});

var _blockEditor = require("@wordpress/block-editor");
