// @wordpress/components/build/color-picker/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _objectSpread3 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classnames = _interopRequireDefault(require("classnames"));

var _lodash = require("lodash");

var _alpha = _interopRequireDefault(require("./alpha"));

var _hue = _interopRequireDefault(require("./hue"));

var _inputs = _interopRequireDefault(require("./inputs"));

var _saturation = _interopRequireDefault(require("./saturation"));

var _utils = require("./utils");

/**
 * Parts of this source were derived and modified from react-color,
 * released under the MIT license.
 *
 * https://github.com/casesandberg/react-color/
 *
 * Copyright (c) 2015 Case Sandberg
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var toLowerCase = function toLowerCase(value) {
  return String(value).toLowerCase();
};

var isValueEmpty = function isValueEmpty(data) {
  if (data.source === 'hex' && !data.hex) {
    return true;
  } else if (data.source === 'hsl' && (!data.h || !data.s || !data.l)) {
    return true;
  } else if (data.source === 'rgb' && (!data.r || !data.g || !data.b) && (!data.h || !data.s || !data.v || !data.a) && (!data.h || !data.s || !data.l || !data.a)) {
    return true;
  }

  return false;
};

var isValidColor = function isValidColor(colors) {
  return colors.hex ? (0, _utils.isValidHex)(colors.hex) : (0, _utils.simpleCheckForValidColor)(colors);
};
/**
 * Function that creates the new color object
 * from old data and the new value.
 *
 * @param {Object} oldColors The old color object.
 * @param {string} oldColors.hex
 * @param {Object} oldColors.rgb
 * @param {number} oldColors.rgb.r
 * @param {number} oldColors.rgb.g
 * @param {number} oldColors.rgb.b
 * @param {number} oldColors.rgb.a
 * @param {Object} oldColors.hsl
 * @param {number} oldColors.hsl.h
 * @param {number} oldColors.hsl.s
 * @param {number} oldColors.hsl.l
 * @param {number} oldColors.hsl.a
 * @param {string} oldColors.draftHex Same format as oldColors.hex
 * @param {Object} oldColors.draftRgb Same format as oldColors.rgb
 * @param {Object} oldColors.draftHsl Same format as oldColors.hsl
 * @param {Object} data Data containing the new value to update.
 * @param {Object} data.source One of `hex`, `rgb`, `hsl`.
 * @param {string|number} data.value Value to update.
 * @param {string} data.valueKey Depends on `data.source` values:
 *   - when source = `rgb`, valuKey can be `r`, `g`, `b`, or `a`.
 *   - when source = `hsl`, valuKey can be `h`, `s`, `l`, or `a`.
 * @return {Object} A new color object for a specific source. For example:
 * { source: 'rgb', r: 1, g: 2, b:3, a:0 }
 */


var dataToColors = function dataToColors(oldColors, _ref) {
  var source = _ref.source,
      valueKey = _ref.valueKey,
      value = _ref.value;

  if (source === 'hex') {
    return (0, _defineProperty2.default)({
      source: source
    }, source, value);
  }

  return (0, _objectSpread3.default)({
    source: source
  }, (0, _objectSpread3.default)({}, oldColors[source], (0, _defineProperty2.default)({}, valueKey, value)));
};

var ColorPicker =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(ColorPicker, _Component);

  function ColorPicker(_ref3) {
    var _this;

    var _ref3$color = _ref3.color,
        color = _ref3$color === void 0 ? '0071a1' : _ref3$color;
    (0, _classCallCheck2.default)(this, ColorPicker);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(ColorPicker).apply(this, arguments));
    var colors = (0, _utils.colorToState)(color);
    _this.state = (0, _objectSpread3.default)({}, colors, {
      draftHex: toLowerCase(colors.hex),
      draftRgb: colors.rgb,
      draftHsl: colors.hsl
    });
    _this.commitValues = _this.commitValues.bind((0, _assertThisInitialized2.default)(_this));
    _this.setDraftValues = _this.setDraftValues.bind((0, _assertThisInitialized2.default)(_this));
    _this.resetDraftValues = _this.resetDraftValues.bind((0, _assertThisInitialized2.default)(_this));
    _this.handleInputChange = _this.handleInputChange.bind((0, _assertThisInitialized2.default)(_this));
    return _this;
  }

  (0, _createClass2.default)(ColorPicker, [{
    key: "commitValues",
    value: function commitValues(data) {
      var _this$props = this.props,
          oldHue = _this$props.oldHue,
          _this$props$onChangeC = _this$props.onChangeComplete,
          onChangeComplete = _this$props$onChangeC === void 0 ? _lodash.noop : _this$props$onChangeC;

      if (isValidColor(data)) {
        var colors = (0, _utils.colorToState)(data, data.h || oldHue);
        this.setState((0, _objectSpread3.default)({}, colors, {
          draftHex: toLowerCase(colors.hex),
          draftHsl: colors.hsl,
          draftRgb: colors.rgb
        }), (0, _lodash.debounce)((0, _lodash.partial)(onChangeComplete, colors), 100));
      }
    }
  }, {
    key: "resetDraftValues",
    value: function resetDraftValues() {
      this.setState({
        draftHex: this.state.hex,
        draftHsl: this.state.hsl,
        draftRgb: this.state.rgb
      });
    }
  }, {
    key: "setDraftValues",
    value: function setDraftValues(data) {
      switch (data.source) {
        case 'hex':
          this.setState({
            draftHex: toLowerCase(data.hex)
          });
          break;

        case 'rgb':
          this.setState({
            draftRgb: data
          });
          break;

        case 'hsl':
          this.setState({
            draftHsl: data
          });
          break;
      }
    }
  }, {
    key: "handleInputChange",
    value: function handleInputChange(data) {
      switch (data.state) {
        case 'reset':
          this.resetDraftValues();
          break;

        case 'commit':
          var colors = dataToColors(this.state, data);

          if (!isValueEmpty(colors)) {
            this.commitValues(colors);
          }

          break;

        case 'draft':
          this.setDraftValues(dataToColors(this.state, data));
          break;
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          className = _this$props2.className,
          disableAlpha = _this$props2.disableAlpha;
      var _this$state = this.state,
          color = _this$state.color,
          hsl = _this$state.hsl,
          hsv = _this$state.hsv,
          rgb = _this$state.rgb,
          draftHex = _this$state.draftHex,
          draftHsl = _this$state.draftHsl,
          draftRgb = _this$state.draftRgb;
      var classes = (0, _classnames.default)(className, {
        'components-color-picker': true,
        'is-alpha-disabled': disableAlpha,
        'is-alpha-enabled': !disableAlpha
      });
      return (0, _element.createElement)("div", {
        className: classes
      }, (0, _element.createElement)("div", {
        className: "components-color-picker__saturation"
      }, (0, _element.createElement)(_saturation.default, {
        hsl: hsl,
        hsv: hsv,
        onChange: this.commitValues
      })), (0, _element.createElement)("div", {
        className: "components-color-picker__body"
      }, (0, _element.createElement)("div", {
        className: "components-color-picker__controls"
      }, (0, _element.createElement)("div", {
        className: "components-color-picker__swatch"
      }, (0, _element.createElement)("div", {
        className: "components-color-picker__active",
        style: {
          backgroundColor: color && color.toRgbString()
        }
      })), (0, _element.createElement)("div", {
        className: "components-color-picker__toggles"
      }, (0, _element.createElement)(_hue.default, {
        hsl: hsl,
        onChange: this.commitValues
      }), disableAlpha ? null : (0, _element.createElement)(_alpha.default, {
        rgb: rgb,
        hsl: hsl,
        onChange: this.commitValues
      }))), (0, _element.createElement)(_inputs.default, {
        rgb: draftRgb,
        hsl: draftHsl,
        hex: draftHex,
        onChange: this.handleInputChange,
        disableAlpha: disableAlpha
      })));
    }
  }]);
  return ColorPicker;
}(_element.Component);

exports.default = ColorPicker;
