// @wordpress/components/build/higher-order/with-fallback-styles/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _lodash = require("lodash");

var _compose = require("@wordpress/compose");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */
var _default = function _default(mapNodeToProps) {
  return (0, _compose.createHigherOrderComponent)(function (WrappedComponent) {
    return (
      /*#__PURE__*/
      function (_Component) {
        (0, _inherits2.default)(_class, _Component);

        function _class() {
          var _this;

          (0, _classCallCheck2.default)(this, _class);
          _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(_class).apply(this, arguments));
          _this.nodeRef = _this.props.node;
          _this.state = {
            fallbackStyles: undefined,
            grabStylesCompleted: false
          };
          _this.bindRef = _this.bindRef.bind((0, _assertThisInitialized2.default)(_this));
          return _this;
        }

        (0, _createClass2.default)(_class, [{
          key: "bindRef",
          value: function bindRef(node) {
            if (!node) {
              return;
            }

            this.nodeRef = node;
          }
        }, {
          key: "componentDidMount",
          value: function componentDidMount() {
            this.grabFallbackStyles();
          }
        }, {
          key: "componentDidUpdate",
          value: function componentDidUpdate() {
            this.grabFallbackStyles();
          }
        }, {
          key: "grabFallbackStyles",
          value: function grabFallbackStyles() {
            var _this$state = this.state,
                grabStylesCompleted = _this$state.grabStylesCompleted,
                fallbackStyles = _this$state.fallbackStyles;

            if (this.nodeRef && !grabStylesCompleted) {
              var newFallbackStyles = mapNodeToProps(this.nodeRef, this.props);

              if (!(0, _lodash.isEqual)(newFallbackStyles, fallbackStyles)) {
                this.setState({
                  fallbackStyles: newFallbackStyles,
                  grabStylesCompleted: !!(0, _lodash.every)(newFallbackStyles)
                });
              }
            }
          }
        }, {
          key: "render",
          value: function render() {
            var wrappedComponent = (0, _element.createElement)(WrappedComponent, (0, _extends2.default)({}, this.props, this.state.fallbackStyles));
            return this.props.node ? wrappedComponent : (0, _element.createElement)("div", {
              ref: this.bindRef
            }, " ", wrappedComponent, " ");
          }
        }]);
        return _class;
      }(_element.Component)
    );
  }, 'withFallbackStyles');
};

exports.default = _default;
