// @wordpress/editor/build-module/components/deprecated.js
import _extends from "@babel/runtime/helpers/esm/extends";
import { createElement } from "@wordpress/element";
// Block Creation Components

/**
 * WordPress dependencies
 */
import deprecated from '@wordpress/deprecated';
import { forwardRef } from '@wordpress/element';
import { Autocomplete as RootAutocomplete, AlignmentToolbar as RootAlignmentToolbar, BlockAlignmentToolbar as RootBlockAlignmentToolbar, BlockControls as RootBlockControls, BlockEdit as RootBlockEdit, BlockEditorKeyboardShortcuts as RootBlockEditorKeyboardShortcuts, BlockFormatControls as RootBlockFormatControls, BlockIcon as RootBlockIcon, BlockInspector as RootBlockInspector, BlockList as RootBlockList, BlockMover as RootBlockMover, BlockNavigationDropdown as RootBlockNavigationDropdown, BlockSelectionClearer as RootBlockSelectionClearer, BlockSettingsMenu as RootBlockSettingsMenu, BlockTitle as RootBlockTitle, BlockToolbar as RootBlockToolbar, ColorPalette as RootColorPalette, ContrastChecker as RootContrastChecker, CopyHandler as RootCopyHandler, createCustomColorsHOC as rootCreateCustomColorsHOC, DefaultBlockAppender as RootDefaultBlockAppender, FontSizePicker as RootFontSizePicker, getColorClassName as rootGetColorClassName, getColorObjectByAttributeValues as rootGetColorObjectByAttributeValues, getColorObjectByColorValue as rootGetColorObjectByColorValue, getFontSize as rootGetFontSize, getFontSizeClass as rootGetFontSizeClass, Inserter as RootInserter, InnerBlocks as RootInnerBlocks, InspectorAdvancedControls as RootInspectorAdvancedControls, InspectorControls as RootInspectorControls, PanelColorSettings as RootPanelColorSettings, PlainText as RootPlainText, RichText as RootRichText, RichTextShortcut as RootRichTextShortcut, RichTextToolbarButton as RootRichTextToolbarButton, __unstableRichTextInputEvent as __unstableRootRichTextInputEvent, MediaPlaceholder as RootMediaPlaceholder, MediaUpload as RootMediaUpload, MediaUploadCheck as RootMediaUploadCheck, MultiBlocksSwitcher as RootMultiBlocksSwitcher, MultiSelectScrollIntoView as RootMultiSelectScrollIntoView, NavigableToolbar as RootNavigableToolbar, ObserveTyping as RootObserveTyping, PreserveScrollInReorder as RootPreserveScrollInReorder, SkipToSelectedBlock as RootSkipToSelectedBlock, URLInput as RootURLInput, URLInputButton as RootURLInputButton, URLPopover as RootURLPopover, Warning as RootWarning, WritingFlow as RootWritingFlow, withColorContext as rootWithColorContext, withColors as rootWithColors, withFontSizes as rootWithFontSizes } from '@wordpress/block-editor';
export { default as ServerSideRender } from '@wordpress/server-side-render';

function deprecateComponent(name, Wrapped) {
  var staticsToHoist = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : [];
  var Component = forwardRef(function (props, ref) {
    deprecated('wp.editor.' + name, {
      alternative: 'wp.blockEditor.' + name
    });
    return createElement(Wrapped, _extends({
      ref: ref
    }, props));
  });
  staticsToHoist.forEach(function (staticName) {
    Component[staticName] = deprecateComponent(name + '.' + staticName, Wrapped[staticName]);
  });
  return Component;
}

function deprecateFunction(name, func) {
  return function () {
    deprecated('wp.editor.' + name, {
      alternative: 'wp.blockEditor.' + name
    });
    return func.apply(void 0, arguments);
  };
}

var RichText = deprecateComponent('RichText', RootRichText, ['Content']);
RichText.isEmpty = deprecateFunction('RichText.isEmpty', RootRichText.isEmpty);
export { RichText };
export var Autocomplete = deprecateComponent('Autocomplete', RootAutocomplete);
export var AlignmentToolbar = deprecateComponent('AlignmentToolbar', RootAlignmentToolbar);
export var BlockAlignmentToolbar = deprecateComponent('BlockAlignmentToolbar', RootBlockAlignmentToolbar);
export var BlockControls = deprecateComponent('BlockControls', RootBlockControls, ['Slot']);
export var BlockEdit = deprecateComponent('BlockEdit', RootBlockEdit);
export var BlockEditorKeyboardShortcuts = deprecateComponent('BlockEditorKeyboardShortcuts', RootBlockEditorKeyboardShortcuts);
export var BlockFormatControls = deprecateComponent('BlockFormatControls', RootBlockFormatControls, ['Slot']);
export var BlockIcon = deprecateComponent('BlockIcon', RootBlockIcon);
export var BlockInspector = deprecateComponent('BlockInspector', RootBlockInspector);
export var BlockList = deprecateComponent('BlockList', RootBlockList);
export var BlockMover = deprecateComponent('BlockMover', RootBlockMover);
export var BlockNavigationDropdown = deprecateComponent('BlockNavigationDropdown', RootBlockNavigationDropdown);
export var BlockSelectionClearer = deprecateComponent('BlockSelectionClearer', RootBlockSelectionClearer);
export var BlockSettingsMenu = deprecateComponent('BlockSettingsMenu', RootBlockSettingsMenu);
export var BlockTitle = deprecateComponent('BlockTitle', RootBlockTitle);
export var BlockToolbar = deprecateComponent('BlockToolbar', RootBlockToolbar);
export var ColorPalette = deprecateComponent('ColorPalette', RootColorPalette);
export var ContrastChecker = deprecateComponent('ContrastChecker', RootContrastChecker);
export var CopyHandler = deprecateComponent('CopyHandler', RootCopyHandler);
export var DefaultBlockAppender = deprecateComponent('DefaultBlockAppender', RootDefaultBlockAppender);
export var FontSizePicker = deprecateComponent('FontSizePicker', RootFontSizePicker);
export var Inserter = deprecateComponent('Inserter', RootInserter);
export var InnerBlocks = deprecateComponent('InnerBlocks', RootInnerBlocks, ['ButtonBlockAppender', 'DefaultBlockAppender', 'Content']);
export var InspectorAdvancedControls = deprecateComponent('InspectorAdvancedControls', RootInspectorAdvancedControls, ['Slot']);
export var InspectorControls = deprecateComponent('InspectorControls', RootInspectorControls, ['Slot']);
export var PanelColorSettings = deprecateComponent('PanelColorSettings', RootPanelColorSettings);
export var PlainText = deprecateComponent('PlainText', RootPlainText);
export var RichTextShortcut = deprecateComponent('RichTextShortcut', RootRichTextShortcut);
export var RichTextToolbarButton = deprecateComponent('RichTextToolbarButton', RootRichTextToolbarButton);
export var __unstableRichTextInputEvent = deprecateComponent('__unstableRichTextInputEvent', __unstableRootRichTextInputEvent);
export var MediaPlaceholder = deprecateComponent('MediaPlaceholder', RootMediaPlaceholder);
export var MediaUpload = deprecateComponent('MediaUpload', RootMediaUpload);
export var MediaUploadCheck = deprecateComponent('MediaUploadCheck', RootMediaUploadCheck);
export var MultiBlocksSwitcher = deprecateComponent('MultiBlocksSwitcher', RootMultiBlocksSwitcher);
export var MultiSelectScrollIntoView = deprecateComponent('MultiSelectScrollIntoView', RootMultiSelectScrollIntoView);
export var NavigableToolbar = deprecateComponent('NavigableToolbar', RootNavigableToolbar);
export var ObserveTyping = deprecateComponent('ObserveTyping', RootObserveTyping);
export var PreserveScrollInReorder = deprecateComponent('PreserveScrollInReorder', RootPreserveScrollInReorder);
export var SkipToSelectedBlock = deprecateComponent('SkipToSelectedBlock', RootSkipToSelectedBlock);
export var URLInput = deprecateComponent('URLInput', RootURLInput);
export var URLInputButton = deprecateComponent('URLInputButton', RootURLInputButton);
export var URLPopover = deprecateComponent('URLPopover', RootURLPopover);
export var Warning = deprecateComponent('Warning', RootWarning);
export var WritingFlow = deprecateComponent('WritingFlow', RootWritingFlow);
export var createCustomColorsHOC = deprecateFunction('createCustomColorsHOC', rootCreateCustomColorsHOC);
export var getColorClassName = deprecateFunction('getColorClassName', rootGetColorClassName);
export var getColorObjectByAttributeValues = deprecateFunction('getColorObjectByAttributeValues', rootGetColorObjectByAttributeValues);
export var getColorObjectByColorValue = deprecateFunction('getColorObjectByColorValue', rootGetColorObjectByColorValue);
export var getFontSize = deprecateFunction('getFontSize', rootGetFontSize);
export var getFontSizeClass = deprecateFunction('getFontSizeClass', rootGetFontSizeClass);
export var withColorContext = deprecateFunction('withColorContext', rootWithColorContext);
export var withColors = deprecateFunction('withColors', rootWithColors);
export var withFontSizes = deprecateFunction('withFontSizes', rootWithFontSizes);
