// @wordpress/block-library/build-module/more/edit.native.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { View } from 'react-native';
import Hr from 'react-native-hr';
/**
 * WordPress dependencies
 */

import { __ } from '@wordpress/i18n';
import { Component } from '@wordpress/element';
/**
 * Internal dependencies
 */

import styles from './editor.scss';

var MoreEdit =
/*#__PURE__*/
function (_Component) {
  _inherits(MoreEdit, _Component);

  function MoreEdit() {
    var _this;

    _classCallCheck(this, MoreEdit);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(MoreEdit).apply(this, arguments));
    _this.state = {
      defaultText: __('Read more')
    };
    return _this;
  }

  _createClass(MoreEdit, [{
    key: "render",
    value: function render() {
      var customText = this.props.attributes.customText;
      var defaultText = this.state.defaultText;
      var content = customText || defaultText;
      return createElement(View, null, createElement(Hr, {
        text: content,
        marginLeft: 0,
        marginRight: 0,
        textStyle: styles['block-library-more__text'],
        lineStyle: styles['block-library-more__line']
      }));
    }
  }]);

  return MoreEdit;
}(Component);

export { MoreEdit as default };
