// @wordpress/block-library/build/social-links/edit.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.SocialLinksEdit = void 0;

var _element = require("@wordpress/element");

var _blockEditor = require("@wordpress/block-editor");

var _socialList = _interopRequireDefault(require("../social-link/social-list"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var ALLOWED_BLOCKS = Object.keys(_socialList.default).map(function (site) {
  return 'core/social-link-' + site;
}); // Template contains the links that show when start.

var TEMPLATE = [['core/social-link-wordpress', {
  url: 'https://wordpress.org'
}], ['core/social-link-facebook'], ['core/social-link-twitter'], ['core/social-link-instagram'], ['core/social-link-linkedin'], ['core/social-link-youtube']];

var SocialLinksEdit = function SocialLinksEdit(_ref) {
  var className = _ref.className;
  return (0, _element.createElement)("div", {
    className: className
  }, (0, _element.createElement)(_blockEditor.InnerBlocks, {
    allowedBlocks: ALLOWED_BLOCKS,
    templateLock: false,
    template: TEMPLATE
  }));
};

exports.SocialLinksEdit = SocialLinksEdit;
var _default = SocialLinksEdit;
exports.default = _default;
