// @wordpress/block-library/build/social-links/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.settings = exports.name = exports.metadata = void 0;

var _i18n = require("@wordpress/i18n");

var _edit = _interopRequireDefault(require("./edit"));

var _save = _interopRequireDefault(require("./save"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var metadata = {
  name: "core/social-links",
  category: "widgets",
  icon: "share",
  attributes: {}
};
exports.metadata = metadata;
var name = metadata.name;
exports.name = name;
var settings = {
  title: (0, _i18n.__)('Social links'),
  description: (0, _i18n.__)('Create a block of links to your social media or external sites'),
  supports: {
    align: ['left', 'center', 'right']
  },
  example: {
    innerBlocks: [{
      name: 'core/social-link-wordpress',
      attributes: {
        url: 'https://wordpress.org'
      }
    }, {
      name: 'core/social-link-facebook',
      attributes: {
        url: 'https://www.facebook.com/WordPress/'
      }
    }, {
      name: 'core/social-link-twitter',
      attributes: {
        url: 'https://twitter.com/WordPress'
      }
    }]
  },
  styles: [{
    name: 'default',
    label: (0, _i18n.__)('Default'),
    isDefault: true
  }, {
    name: 'logos-only',
    label: (0, _i18n.__)('Logos Only')
  }, {
    name: 'pill-shape',
    label: (0, _i18n.__)('Pill Shape')
  }],
  edit: _edit.default,
  save: _save.default
};
exports.settings = settings;
