// @wordpress/block-library/build/navigation-menu/edit.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _blockEditor = require("@wordpress/block-editor");

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _useBlockNavigator2 = _interopRequireDefault(require("./use-block-navigator"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function NavigationMenu(_ref) {
  var attributes = _ref.attributes,
      setAttributes = _ref.setAttributes,
      clientId = _ref.clientId;

  var _useBlockNavigator = (0, _useBlockNavigator2.default)(clientId),
      navigatorToolbarButton = _useBlockNavigator.navigatorToolbarButton,
      navigatorModal = _useBlockNavigator.navigatorModal;

  return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_blockEditor.BlockControls, null, (0, _element.createElement)(_components.Toolbar, null, navigatorToolbarButton)), navigatorModal, (0, _element.createElement)(_blockEditor.InspectorControls, null, (0, _element.createElement)(_components.PanelBody, {
    title: (0, _i18n.__)('Menu Settings')
  }, (0, _element.createElement)(_components.CheckboxControl, {
    value: attributes.automaticallyAdd,
    onChange: function onChange(automaticallyAdd) {
      setAttributes({
        automaticallyAdd: automaticallyAdd
      });
    },
    label: (0, _i18n.__)('Automatically add new pages'),
    help: (0, _i18n.__)('Automatically add new top level pages to this menu.')
  }))), (0, _element.createElement)("div", {
    className: "wp-block-navigation-menu"
  }, (0, _element.createElement)(_blockEditor.InnerBlocks, {
    allowedBlocks: ['core/navigation-menu-item'],
    renderAppender: _blockEditor.InnerBlocks.ButtonBlockAppender
  })));
}

var _default = NavigationMenu;
exports.default = _default;
