// @wordpress/block-library/build/social-link/edit.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _classnames = _interopRequireDefault(require("classnames"));

var _blockEditor = require("@wordpress/block-editor");

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _socialList = require("./social-list");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var SocialLinkEdit = function SocialLinkEdit(_ref) {
  var attributes = _ref.attributes,
      setAttributes = _ref.setAttributes,
      isSelected = _ref.isSelected;
  var url = attributes.url,
      site = attributes.site;

  var _useState = (0, _element.useState)(false),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      showURLPopover = _useState2[0],
      setPopover = _useState2[1];

  var classes = (0, _classnames.default)('wp-social-link', 'wp-social-link-' + site, {
    'wp-social-link__is-incomplete': !url
  }); // Import icon.

  var IconComponent = (0, _socialList.getIconBySite)(site);
  return (0, _element.createElement)(_components.Button, {
    className: classes,
    onClick: function onClick() {
      return setPopover(true);
    }
  }, (0, _element.createElement)(IconComponent, null), isSelected && showURLPopover && (0, _element.createElement)(_blockEditor.URLPopover, {
    onClose: function onClose() {
      return setPopover(false);
    }
  }, (0, _element.createElement)("form", {
    className: "block-editor-url-popover__link-editor",
    onSubmit: function onSubmit(event) {
      event.preventDefault();
      setPopover(false);
    }
  }, (0, _element.createElement)("div", {
    className: "editor-url-input block-editor-url-input"
  }, (0, _element.createElement)("input", {
    type: "text",
    value: url,
    onChange: function onChange(event) {
      return setAttributes({
        url: event.target.value
      });
    },
    placeholder: (0, _i18n.__)('Enter Address')
  })), (0, _element.createElement)(_components.IconButton, {
    icon: "editor-break",
    label: (0, _i18n.__)('Apply'),
    type: "submit"
  }))));
};

var _default = SocialLinkEdit;
exports.default = _default;
