// @wordpress/block-library/build/social-link/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.sites = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _i18n = require("@wordpress/i18n");

var _edit = _interopRequireDefault(require("./edit"));

var _socialList = _interopRequireDefault(require("./social-list"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var commonAttributes = {
  category: 'widgets',
  parent: ['core/social-links'],
  supports: {
    reusable: false,
    html: false
  },
  edit: _edit.default
}; // Create individual blocks out of each site in social-list.js

var sites = Object.keys(_socialList.default).map(function (site) {
  var siteParams = _socialList.default[site];
  return {
    name: 'core/social-link-' + site,
    settings: (0, _objectSpread2.default)({
      title: siteParams.name,
      icon: siteParams.icon,
      description: (0, _i18n.__)('Link to ' + siteParams.name)
    }, commonAttributes, {
      attributes: {
        url: {
          type: 'string'
        },
        site: {
          type: 'string',
          default: site
        }
      }
    })
  };
});
exports.sites = sites;
