// @wordpress/block-library/build/social-link/social-list.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getIconBySite = exports.default = void 0;

var _icons = require("./icons");

/**
 * Internal dependencies
 */
var socialList = {
  fivehundredpx: {
    name: '500px',
    icon: _icons.FivehundredpxIcon
  },
  amazon: {
    name: 'Amazon',
    icon: _icons.AmazonIcon
  },
  bandcamp: {
    name: 'Bandcamp',
    icon: _icons.BandcampIcon
  },
  behance: {
    name: 'Behance',
    icon: _icons.BehanceIcon
  },
  chain: {
    name: 'Link',
    icon: _icons.ChainIcon
  },
  codepen: {
    name: 'CodePen',
    icon: _icons.CodepenIcon
  },
  deviantart: {
    name: 'DeviantArt',
    icon: _icons.DeviantartIcon
  },
  dribbble: {
    name: 'Dribbble',
    icon: _icons.DribbbleIcon
  },
  dropbox: {
    name: 'Dropbox',
    icon: _icons.DropboxIcon
  },
  etsy: {
    name: 'Etsy',
    icon: _icons.EtsyIcon
  },
  facebook: {
    name: 'Facebook',
    icon: _icons.FacebookIcon
  },
  feed: {
    name: 'RSS Feed',
    icon: _icons.FeedIcon
  },
  flickr: {
    name: 'Flickr',
    icon: _icons.FlickrIcon
  },
  foursquare: {
    name: 'Foursquare',
    icon: _icons.FoursquareIcon
  },
  goodreads: {
    name: 'Goodreads',
    icon: _icons.GoodreadsIcon
  },
  google: {
    name: 'Google',
    icon: _icons.GoogleIcon
  },
  github: {
    name: 'Github',
    icon: _icons.GithubIcon
  },
  instagram: {
    name: 'Instagram',
    icon: _icons.InstagramIcon
  },
  lastfm: {
    name: 'Last.fm',
    icon: _icons.LastfmIcon
  },
  linkedin: {
    name: 'Linkedin',
    icon: _icons.LinkedinIcon
  },
  mail: {
    name: 'Mail',
    icon: _icons.MailIcon
  },
  mastodon: {
    name: 'Mastodon',
    icon: _icons.MastodonIcon
  },
  meetup: {
    name: 'Meetup',
    icon: _icons.MeetupIcon
  },
  medium: {
    name: 'Medium',
    icon: _icons.MediumIcon
  },
  pinterest: {
    name: 'Pinterest',
    icon: _icons.PinterestIcon
  },
  pocket: {
    name: 'Pocket',
    icon: _icons.PocketIcon
  },
  reddit: {
    name: 'Reddit',
    icon: _icons.RedditIcon
  },
  skype: {
    name: 'Skype',
    icon: _icons.SkypeIcon
  },
  snapchat: {
    name: 'Snapchat',
    icon: _icons.SnapchatIcon
  },
  soundcloud: {
    name: 'Soundcloud',
    icon: _icons.SoundcloudIcon
  },
  spotify: {
    name: 'Spotify',
    icon: _icons.SpotifyIcon
  },
  tumblr: {
    name: 'Tumblr',
    icon: _icons.TumblrIcon
  },
  twitch: {
    name: 'Twitch',
    icon: _icons.TwitchIcon
  },
  twitter: {
    name: 'Twitter',
    icon: _icons.TwitterIcon
  },
  vimeo: {
    name: 'Vimeo',
    icon: _icons.VimeoIcon
  },
  vk: {
    name: 'VK',
    icon: _icons.VkIcon
  },
  wordpress: {
    name: 'WordPress',
    icon: _icons.WordPressIcon
  },
  yelp: {
    name: 'Yelp',
    icon: _icons.YelpIcon
  },
  youtube: {
    name: 'YouTube',
    icon: _icons.YoutubeIcon
  }
};
var _default = socialList;
exports.default = _default;

var getIconBySite = function getIconBySite(site) {
  return socialList[site].icon;
};

exports.getIconBySite = getIconBySite;
