// @wordpress/block-library/build/code/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.settings = exports.name = exports.metadata = void 0;

var _i18n = require("@wordpress/i18n");

var _edit = _interopRequireDefault(require("./edit"));

var _icon = _interopRequireDefault(require("./icon"));

var _save = _interopRequireDefault(require("./save"));

var _transforms = _interopRequireDefault(require("./transforms"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var metadata = {
  name: "core/code",
  category: "formatting",
  attributes: {
    content: {
      type: "string",
      source: "text",
      selector: "code"
    }
  }
};
exports.metadata = metadata;
var name = metadata.name;
exports.name = name;
var settings = {
  title: (0, _i18n.__)('Code'),
  description: (0, _i18n.__)('Display code snippets that respect your spacing and tabs.'),
  icon: _icon.default,
  example: {
    attributes: {
      content: (0, _i18n.__)('// A "block" is the abstract term used') + '\n' + (0, _i18n.__)('// to describe units of markup that,') + '\n' + (0, _i18n.__)('// when composed together, form the') + '\n' + (0, _i18n.__)('// content or layout of a page.') + '\n' + (0, _i18n.__)('registerBlockType( name, settings );')
    }
  },
  supports: {
    html: false
  },
  transforms: _transforms.default,
  edit: _edit.default,
  save: _save.default
};
exports.settings = settings;
