// @wordpress/block-editor/build/components/block-list/block.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _reactNative = require("react-native");

var _data = require("@wordpress/data");

var _compose = require("@wordpress/compose");

var _blocks = require("@wordpress/blocks");

var _i18n = require("@wordpress/i18n");

var _block = _interopRequireDefault(require("./block.scss"));

var _blockEdit = _interopRequireDefault(require("../block-edit"));

var _blockInvalidWarning = _interopRequireDefault(require("./block-invalid-warning"));

var _blockMobileToolbar = _interopRequireDefault(require("./block-mobile-toolbar"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var BlockListBlock =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(BlockListBlock, _Component);

  function BlockListBlock() {
    var _this;

    (0, _classCallCheck2.default)(this, BlockListBlock);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(BlockListBlock).apply(this, arguments));
    _this.insertBlocksAfter = _this.insertBlocksAfter.bind((0, _assertThisInitialized2.default)(_this));
    _this.onFocus = _this.onFocus.bind((0, _assertThisInitialized2.default)(_this));
    _this.state = {
      isFullyBordered: false
    };
    return _this;
  }

  (0, _createClass2.default)(BlockListBlock, [{
    key: "onFocus",
    value: function onFocus() {
      if (!this.props.isSelected) {
        this.props.onSelect();
      }
    }
  }, {
    key: "insertBlocksAfter",
    value: function insertBlocksAfter(blocks) {
      this.props.onInsertBlocks(blocks, this.props.order + 1);

      if (blocks[0]) {
        // focus on the first block inserted
        this.props.onSelect(blocks[0].clientId);
      }
    }
  }, {
    key: "getBlockForType",
    value: function getBlockForType() {
      return (0, _element.createElement)(_blockEdit.default, {
        name: this.props.name,
        isSelected: this.props.isSelected,
        attributes: this.props.attributes,
        setAttributes: this.props.onChange,
        onFocus: this.onFocus,
        onReplace: this.props.onReplace,
        insertBlocksAfter: this.insertBlocksAfter,
        mergeBlocks: this.props.mergeBlocks,
        onCaretVerticalPositionChange: this.props.onCaretVerticalPositionChange,
        clientId: this.props.clientId
      });
    }
  }, {
    key: "renderBlockTitle",
    value: function renderBlockTitle() {
      return (0, _element.createElement)(_reactNative.View, {
        style: _block.default.blockTitle
      }, (0, _element.createElement)(_reactNative.Text, null, "BlockType: ", this.props.name));
    }
  }, {
    key: "getAccessibilityLabel",
    value: function getAccessibilityLabel() {
      var _this$props = this.props,
          attributes = _this$props.attributes,
          name = _this$props.name,
          order = _this$props.order,
          title = _this$props.title,
          getAccessibilityLabelExtra = _this$props.getAccessibilityLabelExtra;
      var blockName = '';

      if (name === 'core/missing') {
        // is the block unrecognized?
        blockName = title;
      } else {
        blockName = (0, _i18n.sprintf)(
        /* translators: accessibility text. %s: block name. */
        (0, _i18n.__)('%s Block'), title //already localized
        );
      }

      blockName += '. ' + (0, _i18n.sprintf)((0, _i18n.__)('Row %d.'), order + 1);

      if (getAccessibilityLabelExtra) {
        var blockAccessibilityLabel = getAccessibilityLabelExtra(attributes);
        blockName += blockAccessibilityLabel ? ' ' + blockAccessibilityLabel : '';
      }

      return blockName;
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          borderStyle = _this$props2.borderStyle,
          clientId = _this$props2.clientId,
          focusedBorderColor = _this$props2.focusedBorderColor,
          icon = _this$props2.icon,
          isSelected = _this$props2.isSelected,
          isValid = _this$props2.isValid,
          showTitle = _this$props2.showTitle,
          title = _this$props2.title;
      var borderColor = isSelected ? focusedBorderColor : 'transparent';
      var accessibilityLabel = this.getAccessibilityLabel();
      return (0, _element.createElement)(_reactNative.TouchableWithoutFeedback, {
        onPress: this.onFocus,
        accessible: !isSelected,
        accessibilityRole: 'button'
      }, (0, _element.createElement)(_reactNative.View, {
        style: [_block.default.blockHolder, borderStyle, {
          borderColor: borderColor
        }]
      }, showTitle && this.renderBlockTitle(), (0, _element.createElement)(_reactNative.View, {
        accessibilityLabel: accessibilityLabel,
        style: [!isSelected && _block.default.blockContainer, isSelected && _block.default.blockContainerFocused]
      }, isValid && this.getBlockForType(), !isValid && (0, _element.createElement)(_blockInvalidWarning.default, {
        blockTitle: title,
        icon: icon
      })), isSelected && (0, _element.createElement)(_blockMobileToolbar.default, {
        clientId: clientId
      })));
    }
  }]);
  return BlockListBlock;
}(_element.Component);

var _default = (0, _compose.compose)([(0, _data.withSelect)(function (select, _ref) {
  var clientId = _ref.clientId,
      rootClientId = _ref.rootClientId;

  var _select = select('core/block-editor'),
      getBlockIndex = _select.getBlockIndex,
      getBlocks = _select.getBlocks,
      isBlockSelected = _select.isBlockSelected,
      __unstableGetBlockWithoutInnerBlocks = _select.__unstableGetBlockWithoutInnerBlocks;

  var order = getBlockIndex(clientId, rootClientId);
  var isSelected = isBlockSelected(clientId);
  var isFirstBlock = order === 0;
  var isLastBlock = order === getBlocks().length - 1;

  var block = __unstableGetBlockWithoutInnerBlocks(clientId);

  var _ref2 = block || {},
      name = _ref2.name,
      attributes = _ref2.attributes,
      isValid = _ref2.isValid;

  var blockType = (0, _blocks.getBlockType)(name || 'core/missing');
  var title = blockType.title;
  var icon = blockType.icon;
  var getAccessibilityLabelExtra = blockType.__experimentalGetAccessibilityLabel;
  return {
    icon: icon,
    name: name || 'core/missing',
    order: order,
    title: title,
    attributes: attributes,
    blockType: blockType,
    isFirstBlock: isFirstBlock,
    isLastBlock: isLastBlock,
    isSelected: isSelected,
    isValid: isValid,
    getAccessibilityLabelExtra: getAccessibilityLabelExtra
  };
}), (0, _data.withDispatch)(function (dispatch, ownProps, _ref3) {
  var select = _ref3.select;

  var _dispatch = dispatch('core/block-editor'),
      insertBlocks = _dispatch.insertBlocks,
      _mergeBlocks = _dispatch.mergeBlocks,
      replaceBlocks = _dispatch.replaceBlocks,
      selectBlock = _dispatch.selectBlock,
      updateBlockAttributes = _dispatch.updateBlockAttributes;

  return {
    mergeBlocks: function mergeBlocks(forward) {
      var clientId = ownProps.clientId;

      var _select2 = select('core/block-editor'),
          getPreviousBlockClientId = _select2.getPreviousBlockClientId,
          getNextBlockClientId = _select2.getNextBlockClientId;

      if (forward) {
        var nextBlockClientId = getNextBlockClientId(clientId);

        if (nextBlockClientId) {
          _mergeBlocks(clientId, nextBlockClientId);
        }
      } else {
        var previousBlockClientId = getPreviousBlockClientId(clientId);

        if (previousBlockClientId) {
          _mergeBlocks(previousBlockClientId, clientId);
        }
      }
    },
    onInsertBlocks: function onInsertBlocks(blocks, index) {
      insertBlocks(blocks, index, ownProps.rootClientId);
    },
    onSelect: function onSelect() {
      var clientId = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : ownProps.clientId;
      var initialPosition = arguments.length > 1 ? arguments[1] : undefined;
      selectBlock(clientId, initialPosition);
    },
    onChange: function onChange(attributes) {
      updateBlockAttributes(ownProps.clientId, attributes);
    },
    onReplace: function onReplace(blocks, indexToSelect) {
      replaceBlocks([ownProps.clientId], blocks, indexToSelect);
    }
  };
})])(BlockListBlock);

exports.default = _default;
