// @wordpress/components/build/focal-point-picker/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.FocalPointPicker = void 0;

var _element = require("@wordpress/element");

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _classnames = _interopRequireDefault(require("classnames"));

var _i18n = require("@wordpress/i18n");

var _compose = require("@wordpress/compose");

var _baseControl = _interopRequireDefault(require("../base-control"));

var _withFocusOutside = _interopRequireDefault(require("../higher-order/with-focus-outside"));

var _primitives = require("../primitives");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var TEXTCONTROL_MIN = 0;
var TEXTCONTROL_MAX = 100;

var FocalPointPicker =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(FocalPointPicker, _Component);

  function FocalPointPicker(props) {
    var _this;

    (0, _classCallCheck2.default)(this, FocalPointPicker);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(FocalPointPicker).call(this, props));
    _this.onMouseMove = _this.onMouseMove.bind((0, _assertThisInitialized2.default)(_this));
    _this.state = {
      isDragging: false,
      bounds: {},
      percentages: props.value
    };
    _this.containerRef = (0, _element.createRef)();
    _this.imageRef = (0, _element.createRef)();
    _this.horizontalPositionChanged = _this.horizontalPositionChanged.bind((0, _assertThisInitialized2.default)(_this));
    _this.verticalPositionChanged = _this.verticalPositionChanged.bind((0, _assertThisInitialized2.default)(_this));
    _this.onLoad = _this.onLoad.bind((0, _assertThisInitialized2.default)(_this));
    return _this;
  }

  (0, _createClass2.default)(FocalPointPicker, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      if (prevProps.url !== this.props.url) {
        this.setState({
          isDragging: false
        });
      }
    }
  }, {
    key: "calculateBounds",
    value: function calculateBounds() {
      var bounds = {
        top: 0,
        left: 0,
        bottom: 0,
        right: 0,
        width: 0,
        height: 0
      };

      if (!this.imageRef.current) {
        return bounds;
      }

      var dimensions = {
        width: this.imageRef.current.clientWidth,
        height: this.imageRef.current.clientHeight
      };
      var pickerDimensions = this.pickerDimensions();
      var widthRatio = pickerDimensions.width / dimensions.width;
      var heightRatio = pickerDimensions.height / dimensions.height;

      if (heightRatio >= widthRatio) {
        bounds.width = bounds.right = pickerDimensions.width;
        bounds.height = dimensions.height * widthRatio;
        bounds.top = (pickerDimensions.height - bounds.height) / 2;
        bounds.bottom = bounds.top + bounds.height;
      } else {
        bounds.height = bounds.bottom = pickerDimensions.height;
        bounds.width = dimensions.width * heightRatio;
        bounds.left = (pickerDimensions.width - bounds.width) / 2;
        bounds.right = bounds.left + bounds.width;
      }

      return bounds;
    }
  }, {
    key: "onLoad",
    value: function onLoad() {
      this.setState({
        bounds: this.calculateBounds()
      });
    }
  }, {
    key: "onMouseMove",
    value: function onMouseMove(event) {
      var _this$state = this.state,
          isDragging = _this$state.isDragging,
          bounds = _this$state.bounds;
      var onChange = this.props.onChange;

      if (isDragging) {
        var pickerDimensions = this.pickerDimensions();
        var cursorPosition = {
          left: event.pageX - pickerDimensions.left,
          top: event.pageY - pickerDimensions.top
        };
        var left = Math.max(bounds.left, Math.min(cursorPosition.left, bounds.right));
        var top = Math.max(bounds.top, Math.min(cursorPosition.top, bounds.bottom));
        var percentages = {
          x: (left - bounds.left) / (pickerDimensions.width - bounds.left * 2),
          y: (top - bounds.top) / (pickerDimensions.height - bounds.top * 2)
        };
        this.setState({
          percentages: percentages
        }, function () {
          onChange({
            x: this.state.percentages.x,
            y: this.state.percentages.y
          });
        });
      }
    }
  }, {
    key: "fractionToPercentage",
    value: function fractionToPercentage(fraction) {
      return Math.round(fraction * 100);
    }
  }, {
    key: "horizontalPositionChanged",
    value: function horizontalPositionChanged(event) {
      this.positionChangeFromTextControl('x', event.target.value);
    }
  }, {
    key: "verticalPositionChanged",
    value: function verticalPositionChanged(event) {
      this.positionChangeFromTextControl('y', event.target.value);
    }
  }, {
    key: "positionChangeFromTextControl",
    value: function positionChangeFromTextControl(axis, value) {
      var onChange = this.props.onChange;
      var percentages = this.state.percentages;
      var cleanValue = Math.max(Math.min(parseInt(value), 100), 0);
      percentages[axis] = cleanValue ? cleanValue / 100 : 0;
      this.setState({
        percentages: percentages
      }, function () {
        onChange({
          x: this.state.percentages.x,
          y: this.state.percentages.y
        });
      });
    }
  }, {
    key: "pickerDimensions",
    value: function pickerDimensions() {
      if (this.containerRef.current) {
        return {
          width: this.containerRef.current.clientWidth,
          height: this.containerRef.current.clientHeight,
          top: this.containerRef.current.getBoundingClientRect().top + document.body.scrollTop,
          left: this.containerRef.current.getBoundingClientRect().left
        };
      }

      return {
        width: 0,
        height: 0,
        left: 0,
        top: 0
      };
    }
  }, {
    key: "handleFocusOutside",
    value: function handleFocusOutside() {
      this.setState({
        isDragging: false
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          instanceId = _this$props.instanceId,
          url = _this$props.url,
          value = _this$props.value,
          label = _this$props.label,
          help = _this$props.help,
          className = _this$props.className;
      var _this$state2 = this.state,
          bounds = _this$state2.bounds,
          isDragging = _this$state2.isDragging,
          percentages = _this$state2.percentages;
      var pickerDimensions = this.pickerDimensions();
      var iconCoordinates = {
        left: value.x * (pickerDimensions.width - bounds.left * 2) + bounds.left,
        top: value.y * (pickerDimensions.height - bounds.top * 2) + bounds.top
      };
      var iconContainerStyle = {
        left: "".concat(iconCoordinates.left, "px"),
        top: "".concat(iconCoordinates.top, "px")
      };
      var iconContainerClasses = (0, _classnames.default)('components-focal-point-picker__icon_container', isDragging ? 'is-dragging' : null);
      var id = "inspector-focal-point-picker-control-".concat(instanceId);
      var horizontalPositionId = "inspector-focal-point-picker-control-horizontal-position-".concat(instanceId);
      var verticalPositionId = "inspector-focal-point-picker-control-vertical-position-".concat(instanceId);
      return (0, _element.createElement)(_baseControl.default, {
        label: label,
        id: id,
        help: help,
        className: className
      }, (0, _element.createElement)("div", {
        className: "components-focal-point-picker-wrapper"
      }, (0, _element.createElement)("div", {
        className: "components-focal-point-picker",
        onMouseDown: function onMouseDown() {
          return _this2.setState({
            isDragging: true
          });
        },
        onDragStart: function onDragStart() {
          return _this2.setState({
            isDragging: true
          });
        },
        onMouseUp: function onMouseUp() {
          return _this2.setState({
            isDragging: false
          });
        },
        onDrop: function onDrop() {
          return _this2.setState({
            isDragging: false
          });
        },
        onMouseMove: this.onMouseMove,
        ref: this.containerRef,
        role: "button",
        tabIndex: "-1"
      }, (0, _element.createElement)("img", {
        alt: "Dimensions helper",
        onLoad: this.onLoad,
        ref: this.imageRef,
        src: url,
        draggable: "false"
      }), (0, _element.createElement)("div", {
        className: iconContainerClasses,
        style: iconContainerStyle
      }, (0, _element.createElement)(_primitives.SVG, {
        className: "components-focal-point-picker__icon",
        xmlns: "http://www.w3.org/2000/svg",
        viewBox: "0 0 30 30"
      }, (0, _element.createElement)(_primitives.Path, {
        className: "components-focal-point-picker__icon-outline",
        d: "M15 1C7.3 1 1 7.3 1 15s6.3 14 14 14 14-6.3 14-14S22.7 1 15 1zm0 22c-4.4 0-8-3.6-8-8s3.6-8 8-8 8 3.6 8 8-3.6 8-8 8z"
      }), (0, _element.createElement)(_primitives.Path, {
        className: "components-focal-point-picker__icon-fill",
        d: "M15 3C8.4 3 3 8.4 3 15s5.4 12 12 12 12-5.4 12-12S21.6 3 15 3zm0 22C9.5 25 5 20.5 5 15S9.5 5 15 5s10 4.5 10 10-4.5 10-10 10z"
      }))))), (0, _element.createElement)("div", {
        className: "components-focal-point-picker_position-display-container"
      }, (0, _element.createElement)(_baseControl.default, {
        label: (0, _i18n.__)('Horizontal Pos.'),
        id: horizontalPositionId
      }, (0, _element.createElement)("input", {
        className: "components-text-control__input",
        id: horizontalPositionId,
        max: TEXTCONTROL_MAX,
        min: TEXTCONTROL_MIN,
        onChange: this.horizontalPositionChanged,
        type: "number",
        value: this.fractionToPercentage(percentages.x)
      }), (0, _element.createElement)("span", null, "%")), (0, _element.createElement)(_baseControl.default, {
        label: (0, _i18n.__)('Vertical Pos.'),
        id: verticalPositionId
      }, (0, _element.createElement)("input", {
        className: "components-text-control__input",
        id: verticalPositionId,
        max: TEXTCONTROL_MAX,
        min: TEXTCONTROL_MIN,
        onChange: this.verticalPositionChanged,
        type: "number",
        value: this.fractionToPercentage(percentages.y)
      }), (0, _element.createElement)("span", null, "%"))));
    }
  }]);
  return FocalPointPicker;
}(_element.Component);

exports.FocalPointPicker = FocalPointPicker;
FocalPointPicker.defaultProps = {
  url: null,
  value: {
    x: 0.5,
    y: 0.5
  },
  onChange: function onChange() {}
};

var _default = (0, _compose.compose)([_compose.withInstanceId, _withFocusOutside.default])(FocalPointPicker);

exports.default = _default;
