// @wordpress/block-directory/build/components/downloadable-block-header/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _blockEditor = require("@wordpress/block-editor");

var _blockRatings = _interopRequireDefault(require("../block-ratings"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function DownloadableBlockHeader(_ref) {
  var icon = _ref.icon,
      title = _ref.title,
      rating = _ref.rating,
      ratingCount = _ref.ratingCount,
      _onClick = _ref.onClick;
  return (0, _element.createElement)("div", {
    className: "block-directory-downloadable-block-header__row"
  }, icon.match(/\.(jpeg|jpg|gif|png)$/) !== null ? (0, _element.createElement)("img", {
    src: icon,
    alt: "block icon"
  }) : (0, _element.createElement)("span", null, (0, _element.createElement)(_blockEditor.BlockIcon, {
    icon: icon,
    showColors: true
  })), (0, _element.createElement)("div", {
    className: "block-directory-downloadable-block-header__column"
  }, (0, _element.createElement)("span", {
    role: "heading",
    className: "block-directory-downloadable-block-header__title"
  }, title), (0, _element.createElement)(_blockRatings.default, {
    rating: rating,
    ratingCount: ratingCount
  })), (0, _element.createElement)(_components.Button, {
    isDefault: true,
    onClick: function onClick(event) {
      event.preventDefault();

      _onClick();
    }
  }, (0, _i18n.__)('Add')));
}

var _default = DownloadableBlockHeader;
exports.default = _default;
