// @wordpress/components/build/mobile/keyboard-aware-flat-list/index.ios.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.KeyboardAwareFlatList = void 0;

var _element = require("@wordpress/element");

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _reactNativeKeyboardAwareScrollView = require("react-native-keyboard-aware-scroll-view");

var _reactNative = require("react-native");

var _this = void 0;

var KeyboardAwareFlatList = function KeyboardAwareFlatList(_ref) {
  var extraScrollHeight = _ref.extraScrollHeight,
      shouldPreventAutomaticScroll = _ref.shouldPreventAutomaticScroll,
      _innerRef = _ref.innerRef,
      listProps = (0, _objectWithoutProperties2.default)(_ref, ["extraScrollHeight", "shouldPreventAutomaticScroll", "innerRef"]);
  return (0, _element.createElement)(_reactNativeKeyboardAwareScrollView.KeyboardAwareScrollView, {
    style: {
      flex: 1
    },
    keyboardDismissMode: "none",
    enableResetScrollToCoords: false,
    keyboardShouldPersistTaps: "handled",
    extraScrollHeight: extraScrollHeight,
    extraHeight: 0,
    innerRef: function innerRef(ref) {
      _this.scrollViewRef = ref;

      _innerRef(ref);
    },
    onKeyboardWillHide: function onKeyboardWillHide() {
      _this.keyboardWillShowIndicator = false;
    },
    onKeyboardDidHide: function onKeyboardDidHide() {
      setTimeout(function () {
        if (!_this.keyboardWillShowIndicator && _this.latestContentOffsetY !== undefined && !shouldPreventAutomaticScroll()) {
          // Reset the content position if keyboard is still closed
          _this.scrollViewRef.props.scrollToPosition(0, _this.latestContentOffsetY, true);
        }
      }, 50);
    },
    onKeyboardWillShow: function onKeyboardWillShow() {
      _this.keyboardWillShowIndicator = true;
    },
    onScroll: function onScroll(event) {
      _this.latestContentOffsetY = event.nativeEvent.contentOffset.y;
    }
  }, (0, _element.createElement)(_reactNative.FlatList, listProps));
};

exports.KeyboardAwareFlatList = KeyboardAwareFlatList;

KeyboardAwareFlatList.handleCaretVerticalPositionChange = function (scrollView, targetId, caretY, previousCaretY) {
  if (previousCaretY) {
    //if this is not the first tap
    scrollView.props.refreshScrollForField(targetId);
  }
};

var _default = KeyboardAwareFlatList;
exports.default = _default;
