// @wordpress/block-editor/build-module/components/block-mover/index.native.js
import { createElement, Fragment } from "@wordpress/element";

/**
 * External dependencies
 */
import { first, last, partial, castArray } from 'lodash';
/**
 * WordPress dependencies
 */

import { ToolbarButton } from '@wordpress/components';
import { __, sprintf } from '@wordpress/i18n';
import { withSelect, withDispatch } from '@wordpress/data';
import { withInstanceId, compose } from '@wordpress/compose';

var BlockMover = function BlockMover(_ref) {
  var isFirst = _ref.isFirst,
      isLast = _ref.isLast,
      onMoveDown = _ref.onMoveDown,
      onMoveUp = _ref.onMoveUp,
      firstIndex = _ref.firstIndex;
  return createElement(Fragment, null, createElement(ToolbarButton, {
    title: !isFirst ? sprintf(
    /* translators: accessibility text. %1: current block position (number). %2: next block position (number) */
    __('Move block up from row %1$s to row %2$s'), firstIndex + 1, firstIndex) : __('Move block up'),
    isDisabled: isFirst,
    onClick: onMoveUp,
    icon: "arrow-up-alt",
    extraProps: {
      hint: __('Double tap to move the block up')
    }
  }), createElement(ToolbarButton, {
    title: !isLast ? sprintf(
    /* translators: accessibility text. %1: current block position (number). %2: next block position (number) */
    __('Move block down from row %1$s to row %2$s'), firstIndex + 1, firstIndex + 2) : __('Move block down'),
    isDisabled: isLast,
    onClick: onMoveDown,
    icon: "arrow-down-alt",
    extraProps: {
      hint: __('Double tap to move the block down')
    }
  }));
};

export default compose(withSelect(function (select, _ref2) {
  var clientIds = _ref2.clientIds;

  var _select = select('core/block-editor'),
      getBlockIndex = _select.getBlockIndex,
      getBlockRootClientId = _select.getBlockRootClientId,
      getBlockOrder = _select.getBlockOrder;

  var normalizedClientIds = castArray(clientIds);
  var firstClientId = first(normalizedClientIds);
  var rootClientId = getBlockRootClientId(first(normalizedClientIds));
  var blockOrder = getBlockOrder(rootClientId);
  var firstIndex = getBlockIndex(firstClientId, rootClientId);
  var lastIndex = getBlockIndex(last(normalizedClientIds), rootClientId);
  return {
    firstIndex: firstIndex,
    isFirst: firstIndex === 0,
    isLast: lastIndex === blockOrder.length - 1
  };
}), withDispatch(function (dispatch, _ref3) {
  var clientIds = _ref3.clientIds,
      rootClientId = _ref3.rootClientId;

  var _dispatch = dispatch('core/block-editor'),
      moveBlocksDown = _dispatch.moveBlocksDown,
      moveBlocksUp = _dispatch.moveBlocksUp;

  return {
    onMoveDown: partial(moveBlocksDown, clientIds, rootClientId),
    onMoveUp: partial(moveBlocksUp, clientIds, rootClientId)
  };
}), withInstanceId)(BlockMover);
