// @wordpress/components/build/higher-order/with-notices/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _v = _interopRequireDefault(require("uuid/v4"));

var _compose = require("@wordpress/compose");

var _list = _interopRequireDefault(require("../../notice/list"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

/**
 * Override the default edit UI to include notices if supported.
 *
 * @param  {Function|Component} OriginalComponent Original component.
 * @return {Component}                            Wrapped component.
 */
var _default = (0, _compose.createHigherOrderComponent)(function (OriginalComponent) {
  return (
    /*#__PURE__*/
    function (_Component) {
      (0, _inherits2.default)(WrappedBlockEdit, _Component);

      function WrappedBlockEdit() {
        var _this;

        (0, _classCallCheck2.default)(this, WrappedBlockEdit);
        _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(WrappedBlockEdit).apply(this, arguments));
        _this.createNotice = _this.createNotice.bind((0, _assertThisInitialized2.default)(_this));
        _this.createErrorNotice = _this.createErrorNotice.bind((0, _assertThisInitialized2.default)(_this));
        _this.removeNotice = _this.removeNotice.bind((0, _assertThisInitialized2.default)(_this));
        _this.removeAllNotices = _this.removeAllNotices.bind((0, _assertThisInitialized2.default)(_this));
        _this.state = {
          noticeList: []
        };
        _this.noticeOperations = {
          createNotice: _this.createNotice,
          createErrorNotice: _this.createErrorNotice,
          removeAllNotices: _this.removeAllNotices,
          removeNotice: _this.removeNotice
        };
        return _this;
      }
      /**
       * Function passed down as a prop that adds a new notice.
       *
       * @param {Object} notice  Notice to add.
       */


      (0, _createClass2.default)(WrappedBlockEdit, [{
        key: "createNotice",
        value: function createNotice(notice) {
          var noticeToAdd = notice.id ? notice : (0, _objectSpread2.default)({}, notice, {
            id: (0, _v.default)()
          });
          this.setState(function (state) {
            return {
              noticeList: [].concat((0, _toConsumableArray2.default)(state.noticeList), [noticeToAdd])
            };
          });
        }
        /**
         * Function passed as a prop that adds a new error notice.
         *
         * @param {string} msg  Error message of the notice.
         */

      }, {
        key: "createErrorNotice",
        value: function createErrorNotice(msg) {
          this.createNotice({
            status: 'error',
            content: msg
          });
        }
        /**
         * Removes a notice by id.
         *
         * @param {string} id  Id of the notice to remove.
         */

      }, {
        key: "removeNotice",
        value: function removeNotice(id) {
          this.setState(function (state) {
            return {
              noticeList: state.noticeList.filter(function (notice) {
                return notice.id !== id;
              })
            };
          });
        }
        /**
         * Removes all notices
         */

      }, {
        key: "removeAllNotices",
        value: function removeAllNotices() {
          this.setState({
            noticeList: []
          });
        }
      }, {
        key: "render",
        value: function render() {
          return (0, _element.createElement)(OriginalComponent, (0, _extends2.default)({
            noticeList: this.state.noticeList,
            noticeOperations: this.noticeOperations,
            noticeUI: this.state.noticeList.length > 0 && (0, _element.createElement)(_list.default, {
              className: "components-with-notices-ui",
              notices: this.state.noticeList,
              onRemove: this.removeNotice
            })
          }, this.props));
        }
      }]);
      return WrappedBlockEdit;
    }(_element.Component)
  );
});

exports.default = _default;
