// @wordpress/block-editor/build-module/components/caption/index.native.js
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { View } from 'react-native';
/**
 * WordPress dependencies
 */

import { RichText } from '@wordpress/block-editor';
import { compose } from '@wordpress/compose';
import { withDispatch, withSelect } from '@wordpress/data';
import { __ } from '@wordpress/i18n';

var Caption = function Caption(_ref) {
  var accessible = _ref.accessible,
      accessibilityLabel = _ref.accessibilityLabel,
      onBlur = _ref.onBlur,
      onChange = _ref.onChange,
      onFocus = _ref.onFocus,
      isSelected = _ref.isSelected,
      shouldDisplay = _ref.shouldDisplay,
      text = _ref.text;
  return createElement(View, {
    style: {
      padding: 12,
      flex: 1,
      display: shouldDisplay ? 'flex' : 'none'
    },
    accessible: accessible,
    accessibilityLabel: accessibilityLabel,
    accessibilityRole: 'button'
  }, createElement(RichText, {
    rootTagsToEliminate: ['p'],
    placeholder: __('Write caption…'),
    value: text,
    onChange: onChange,
    unstableOnFocus: onFocus,
    onBlur: onBlur // always assign onBlur as props
    ,
    isSelected: isSelected,
    __unstableMobileNoFocusOnMount: true,
    fontSize: 14,
    underlineColorAndroid: "transparent",
    textAlign: 'center',
    tagName: ''
  }));
};

export default compose([withSelect(function (select, _ref2) {
  var accessibilityLabelCreator = _ref2.accessibilityLabelCreator,
      clientId = _ref2.clientId;

  var _select = select('core/block-editor'),
      getBlockAttributes = _select.getBlockAttributes,
      getSelectedBlockClientId = _select.getSelectedBlockClientId;

  var _getBlockAttributes = getBlockAttributes(clientId),
      caption = _getBlockAttributes.caption;

  var accessibilityLabel = accessibilityLabelCreator ? accessibilityLabelCreator(caption) : undefined;
  var isBlockSelected = getSelectedBlockClientId() === clientId; // We'll render the caption so that the soft keyboard is not forced to close on Android
  // but still hide it by setting its display style to none. See wordpress-mobile/gutenberg-mobile#1221

  var shouldDisplay = !RichText.isEmpty(caption) > 0 || isBlockSelected;
  return {
    accessibilityLabel: accessibilityLabel,
    shouldDisplay: shouldDisplay,
    text: caption
  };
}), withDispatch(function (dispatch, _ref3) {
  var clientId = _ref3.clientId;

  var _dispatch = dispatch('core/block-editor'),
      updateBlockAttributes = _dispatch.updateBlockAttributes;

  return {
    onChange: function onChange(caption) {
      updateBlockAttributes(clientId, {
        caption: caption
      });
    }
  };
})])(Caption);
