// @wordpress/components/build/disabled/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _lodash = require("lodash");

var _classnames = _interopRequireDefault(require("classnames"));

var _dom = require("@wordpress/dom");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */
var _createContext = (0, _element.createContext)(false),
    Consumer = _createContext.Consumer,
    Provider = _createContext.Provider;
/**
 * Names of control nodes which qualify for disabled behavior.
 *
 * See WHATWG HTML Standard: 4.10.18.5: "Enabling and disabling form controls: the disabled attribute".
 *
 * @see https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#enabling-and-disabling-form-controls:-the-disabled-attribute
 *
 * @type {string[]}
 */


var DISABLED_ELIGIBLE_NODE_NAMES = ['BUTTON', 'FIELDSET', 'INPUT', 'OPTGROUP', 'OPTION', 'SELECT', 'TEXTAREA'];

var Disabled =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(Disabled, _Component);

  function Disabled() {
    var _this;

    (0, _classCallCheck2.default)(this, Disabled);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Disabled).apply(this, arguments));
    _this.bindNode = _this.bindNode.bind((0, _assertThisInitialized2.default)(_this));
    _this.disable = _this.disable.bind((0, _assertThisInitialized2.default)(_this)); // Debounce re-disable since disabling process itself will incur
    // additional mutations which should be ignored.

    _this.debouncedDisable = (0, _lodash.debounce)(_this.disable, {
      leading: true
    });
    return _this;
  }

  (0, _createClass2.default)(Disabled, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.disable();
      this.observer = new window.MutationObserver(this.debouncedDisable);
      this.observer.observe(this.node, {
        childList: true,
        attributes: true,
        subtree: true
      });
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.observer.disconnect();
      this.debouncedDisable.cancel();
    }
  }, {
    key: "bindNode",
    value: function bindNode(node) {
      this.node = node;
    }
  }, {
    key: "disable",
    value: function disable() {
      _dom.focus.focusable.find(this.node).forEach(function (focusable) {
        if ((0, _lodash.includes)(DISABLED_ELIGIBLE_NODE_NAMES, focusable.nodeName)) {
          focusable.setAttribute('disabled', '');
        }

        if (focusable.hasAttribute('tabindex')) {
          focusable.removeAttribute('tabindex');
        }

        if (focusable.hasAttribute('contenteditable')) {
          focusable.setAttribute('contenteditable', 'false');
        }
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          className = _this$props.className,
          props = (0, _objectWithoutProperties2.default)(_this$props, ["className"]);
      return (0, _element.createElement)(Provider, {
        value: true
      }, (0, _element.createElement)("div", (0, _extends2.default)({
        ref: this.bindNode,
        className: (0, _classnames.default)(className, 'components-disabled')
      }, props), this.props.children));
    }
  }]);
  return Disabled;
}(_element.Component);

Disabled.Consumer = Consumer;
var _default = Disabled;
exports.default = _default;
