// @wordpress/block-library/build/more/edit.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _reactNative = require("react-native");

var _reactNativeHr = _interopRequireDefault(require("react-native-hr"));

var _i18n = require("@wordpress/i18n");

var _editor = _interopRequireDefault(require("./editor.scss"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var MoreEdit =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(MoreEdit, _Component);

  function MoreEdit() {
    var _this;

    (0, _classCallCheck2.default)(this, MoreEdit);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(MoreEdit).apply(this, arguments));
    _this.state = {
      defaultText: (0, _i18n.__)('Read more')
    };
    return _this;
  }

  (0, _createClass2.default)(MoreEdit, [{
    key: "render",
    value: function render() {
      var customText = this.props.attributes.customText;
      var defaultText = this.state.defaultText;
      var content = customText || defaultText;
      return (0, _element.createElement)(_reactNative.View, null, (0, _element.createElement)(_reactNativeHr.default, {
        text: content,
        marginLeft: 0,
        marginRight: 0,
        textStyle: _editor.default['block-library-more__text'],
        lineStyle: _editor.default['block-library-more__line']
      }));
    }
  }]);
  return MoreEdit;
}(_element.Component);

exports.default = MoreEdit;
