// @wordpress/block-editor/build/components/index.native.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  BlockControls: true,
  BlockEdit: true,
  BlockFormatControls: true,
  AlignmentToolbar: true,
  InspectorControls: true,
  PlainText: true,
  RichText: true,
  RichTextShortcut: true,
  RichTextToolbarButton: true,
  __unstableRichTextInputEvent: true,
  MediaPlaceholder: true,
  MediaUpload: true,
  MEDIA_TYPE_IMAGE: true,
  MEDIA_TYPE_VIDEO: true,
  URLInput: true,
  BlockInvalidWarning: true,
  Caption: true,
  BlockList: true,
  BlockMover: true,
  BlockToolbar: true,
  DefaultBlockAppender: true,
  Inserter: true,
  BlockEditorProvider: true
};
Object.defineProperty(exports, "BlockControls", {
  enumerable: true,
  get: function get() {
    return _blockControls.default;
  }
});
Object.defineProperty(exports, "BlockEdit", {
  enumerable: true,
  get: function get() {
    return _blockEdit.default;
  }
});
Object.defineProperty(exports, "BlockFormatControls", {
  enumerable: true,
  get: function get() {
    return _blockFormatControls.default;
  }
});
Object.defineProperty(exports, "AlignmentToolbar", {
  enumerable: true,
  get: function get() {
    return _alignmentToolbar.default;
  }
});
Object.defineProperty(exports, "InspectorControls", {
  enumerable: true,
  get: function get() {
    return _inspectorControls.default;
  }
});
Object.defineProperty(exports, "PlainText", {
  enumerable: true,
  get: function get() {
    return _plainText.default;
  }
});
Object.defineProperty(exports, "RichText", {
  enumerable: true,
  get: function get() {
    return _richText.default;
  }
});
Object.defineProperty(exports, "RichTextShortcut", {
  enumerable: true,
  get: function get() {
    return _richText.RichTextShortcut;
  }
});
Object.defineProperty(exports, "RichTextToolbarButton", {
  enumerable: true,
  get: function get() {
    return _richText.RichTextToolbarButton;
  }
});
Object.defineProperty(exports, "__unstableRichTextInputEvent", {
  enumerable: true,
  get: function get() {
    return _richText.__unstableRichTextInputEvent;
  }
});
Object.defineProperty(exports, "MediaPlaceholder", {
  enumerable: true,
  get: function get() {
    return _mediaPlaceholder.default;
  }
});
Object.defineProperty(exports, "MediaUpload", {
  enumerable: true,
  get: function get() {
    return _mediaUpload.default;
  }
});
Object.defineProperty(exports, "MEDIA_TYPE_IMAGE", {
  enumerable: true,
  get: function get() {
    return _mediaUpload.MEDIA_TYPE_IMAGE;
  }
});
Object.defineProperty(exports, "MEDIA_TYPE_VIDEO", {
  enumerable: true,
  get: function get() {
    return _mediaUpload.MEDIA_TYPE_VIDEO;
  }
});
Object.defineProperty(exports, "URLInput", {
  enumerable: true,
  get: function get() {
    return _urlInput.default;
  }
});
Object.defineProperty(exports, "BlockInvalidWarning", {
  enumerable: true,
  get: function get() {
    return _blockInvalidWarning.default;
  }
});
Object.defineProperty(exports, "Caption", {
  enumerable: true,
  get: function get() {
    return _caption.default;
  }
});
Object.defineProperty(exports, "BlockList", {
  enumerable: true,
  get: function get() {
    return _blockList.default;
  }
});
Object.defineProperty(exports, "BlockMover", {
  enumerable: true,
  get: function get() {
    return _blockMover.default;
  }
});
Object.defineProperty(exports, "BlockToolbar", {
  enumerable: true,
  get: function get() {
    return _blockToolbar.default;
  }
});
Object.defineProperty(exports, "DefaultBlockAppender", {
  enumerable: true,
  get: function get() {
    return _defaultBlockAppender.default;
  }
});
Object.defineProperty(exports, "Inserter", {
  enumerable: true,
  get: function get() {
    return _inserter.default;
  }
});
Object.defineProperty(exports, "BlockEditorProvider", {
  enumerable: true,
  get: function get() {
    return _provider.default;
  }
});

var _blockControls = _interopRequireDefault(require("./block-controls"));

var _blockEdit = _interopRequireDefault(require("./block-edit"));

var _blockFormatControls = _interopRequireDefault(require("./block-format-controls"));

var _colors = require("./colors");

Object.keys(_colors).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _colors[key];
    }
  });
});

var _fontSizes = require("./font-sizes");

Object.keys(_fontSizes).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _fontSizes[key];
    }
  });
});

var _alignmentToolbar = _interopRequireDefault(require("./alignment-toolbar"));

var _inspectorControls = _interopRequireDefault(require("./inspector-controls"));

var _plainText = _interopRequireDefault(require("./plain-text"));

var _richText = _interopRequireWildcard(require("./rich-text"));

var _mediaPlaceholder = _interopRequireDefault(require("./media-placeholder"));

var _mediaUpload = _interopRequireWildcard(require("./media-upload"));

var _urlInput = _interopRequireDefault(require("./url-input"));

var _blockInvalidWarning = _interopRequireDefault(require("./block-list/block-invalid-warning"));

var _caption = _interopRequireDefault(require("./caption"));

var _blockList = _interopRequireDefault(require("./block-list"));

var _blockMover = _interopRequireDefault(require("./block-mover"));

var _blockToolbar = _interopRequireDefault(require("./block-toolbar"));

var _defaultBlockAppender = _interopRequireDefault(require("./default-block-appender"));

var _inserter = _interopRequireDefault(require("./inserter"));

var _provider = _interopRequireDefault(require("./provider"));
