// @wordpress/block-editor/build/components/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  AlignmentToolbar: true,
  Autocomplete: true,
  BlockAlignmentToolbar: true,
  BlockControls: true,
  BlockEdit: true,
  BlockFormatControls: true,
  BlockIcon: true,
  BlockNavigationDropdown: true,
  __experimentalBlockNavigationList: true,
  BlockVerticalAlignmentToolbar: true,
  ButtonBlockerAppender: true,
  ColorPalette: true,
  ContrastChecker: true,
  InnerBlocks: true,
  InspectorAdvancedControls: true,
  InspectorControls: true,
  MediaPlaceholder: true,
  MediaUpload: true,
  MediaUploadCheck: true,
  PanelColorSettings: true,
  PlainText: true,
  RichText: true,
  RichTextShortcut: true,
  RichTextToolbarButton: true,
  __unstableRichTextInputEvent: true,
  URLInput: true,
  URLInputButton: true,
  URLPopover: true,
  withColorContext: true,
  __experimentalBlockSettingsMenuFirstItem: true,
  __experimentalBlockSettingsMenuPluginsExtension: true,
  __experimentalInserterMenuExtension: true,
  BlockEditorKeyboardShortcuts: true,
  BlockInspector: true,
  BlockList: true,
  BlockMover: true,
  BlockPreview: true,
  BlockSelectionClearer: true,
  BlockSettingsMenu: true,
  BlockTitle: true,
  BlockToolbar: true,
  CopyHandler: true,
  DefaultBlockAppender: true,
  Inserter: true,
  MultiBlocksSwitcher: true,
  MultiSelectScrollIntoView: true,
  NavigableToolbar: true,
  ObserveTyping: true,
  PreserveScrollInReorder: true,
  SkipToSelectedBlock: true,
  Typewriter: true,
  Warning: true,
  WritingFlow: true,
  BlockEditorProvider: true
};
Object.defineProperty(exports, "AlignmentToolbar", {
  enumerable: true,
  get: function get() {
    return _alignmentToolbar.default;
  }
});
Object.defineProperty(exports, "Autocomplete", {
  enumerable: true,
  get: function get() {
    return _autocomplete.default;
  }
});
Object.defineProperty(exports, "BlockAlignmentToolbar", {
  enumerable: true,
  get: function get() {
    return _blockAlignmentToolbar.default;
  }
});
Object.defineProperty(exports, "BlockControls", {
  enumerable: true,
  get: function get() {
    return _blockControls.default;
  }
});
Object.defineProperty(exports, "BlockEdit", {
  enumerable: true,
  get: function get() {
    return _blockEdit.default;
  }
});
Object.defineProperty(exports, "BlockFormatControls", {
  enumerable: true,
  get: function get() {
    return _blockFormatControls.default;
  }
});
Object.defineProperty(exports, "BlockIcon", {
  enumerable: true,
  get: function get() {
    return _blockIcon.default;
  }
});
Object.defineProperty(exports, "BlockNavigationDropdown", {
  enumerable: true,
  get: function get() {
    return _dropdown.default;
  }
});
Object.defineProperty(exports, "__experimentalBlockNavigationList", {
  enumerable: true,
  get: function get() {
    return _list.default;
  }
});
Object.defineProperty(exports, "BlockVerticalAlignmentToolbar", {
  enumerable: true,
  get: function get() {
    return _blockVerticalAlignmentToolbar.default;
  }
});
Object.defineProperty(exports, "ButtonBlockerAppender", {
  enumerable: true,
  get: function get() {
    return _buttonBlockAppender.default;
  }
});
Object.defineProperty(exports, "ColorPalette", {
  enumerable: true,
  get: function get() {
    return _colorPalette.default;
  }
});
Object.defineProperty(exports, "ContrastChecker", {
  enumerable: true,
  get: function get() {
    return _contrastChecker.default;
  }
});
Object.defineProperty(exports, "InnerBlocks", {
  enumerable: true,
  get: function get() {
    return _innerBlocks.default;
  }
});
Object.defineProperty(exports, "InspectorAdvancedControls", {
  enumerable: true,
  get: function get() {
    return _inspectorAdvancedControls.default;
  }
});
Object.defineProperty(exports, "InspectorControls", {
  enumerable: true,
  get: function get() {
    return _inspectorControls.default;
  }
});
Object.defineProperty(exports, "MediaPlaceholder", {
  enumerable: true,
  get: function get() {
    return _mediaPlaceholder.default;
  }
});
Object.defineProperty(exports, "MediaUpload", {
  enumerable: true,
  get: function get() {
    return _mediaUpload.default;
  }
});
Object.defineProperty(exports, "MediaUploadCheck", {
  enumerable: true,
  get: function get() {
    return _check.default;
  }
});
Object.defineProperty(exports, "PanelColorSettings", {
  enumerable: true,
  get: function get() {
    return _panelColorSettings.default;
  }
});
Object.defineProperty(exports, "PlainText", {
  enumerable: true,
  get: function get() {
    return _plainText.default;
  }
});
Object.defineProperty(exports, "RichText", {
  enumerable: true,
  get: function get() {
    return _richText.default;
  }
});
Object.defineProperty(exports, "RichTextShortcut", {
  enumerable: true,
  get: function get() {
    return _richText.RichTextShortcut;
  }
});
Object.defineProperty(exports, "RichTextToolbarButton", {
  enumerable: true,
  get: function get() {
    return _richText.RichTextToolbarButton;
  }
});
Object.defineProperty(exports, "__unstableRichTextInputEvent", {
  enumerable: true,
  get: function get() {
    return _richText.__unstableRichTextInputEvent;
  }
});
Object.defineProperty(exports, "URLInput", {
  enumerable: true,
  get: function get() {
    return _urlInput.default;
  }
});
Object.defineProperty(exports, "URLInputButton", {
  enumerable: true,
  get: function get() {
    return _button.default;
  }
});
Object.defineProperty(exports, "URLPopover", {
  enumerable: true,
  get: function get() {
    return _urlPopover.default;
  }
});
Object.defineProperty(exports, "withColorContext", {
  enumerable: true,
  get: function get() {
    return _withColorContext.default;
  }
});
Object.defineProperty(exports, "__experimentalBlockSettingsMenuFirstItem", {
  enumerable: true,
  get: function get() {
    return _blockSettingsMenuFirstItem.default;
  }
});
Object.defineProperty(exports, "__experimentalBlockSettingsMenuPluginsExtension", {
  enumerable: true,
  get: function get() {
    return _blockSettingsMenuPluginsExtension.default;
  }
});
Object.defineProperty(exports, "__experimentalInserterMenuExtension", {
  enumerable: true,
  get: function get() {
    return _inserterMenuExtension.default;
  }
});
Object.defineProperty(exports, "BlockEditorKeyboardShortcuts", {
  enumerable: true,
  get: function get() {
    return _blockEditorKeyboardShortcuts.default;
  }
});
Object.defineProperty(exports, "BlockInspector", {
  enumerable: true,
  get: function get() {
    return _blockInspector.default;
  }
});
Object.defineProperty(exports, "BlockList", {
  enumerable: true,
  get: function get() {
    return _blockList.default;
  }
});
Object.defineProperty(exports, "BlockMover", {
  enumerable: true,
  get: function get() {
    return _blockMover.default;
  }
});
Object.defineProperty(exports, "BlockPreview", {
  enumerable: true,
  get: function get() {
    return _blockPreview.default;
  }
});
Object.defineProperty(exports, "BlockSelectionClearer", {
  enumerable: true,
  get: function get() {
    return _blockSelectionClearer.default;
  }
});
Object.defineProperty(exports, "BlockSettingsMenu", {
  enumerable: true,
  get: function get() {
    return _blockSettingsMenu.default;
  }
});
Object.defineProperty(exports, "BlockTitle", {
  enumerable: true,
  get: function get() {
    return _blockTitle.default;
  }
});
Object.defineProperty(exports, "BlockToolbar", {
  enumerable: true,
  get: function get() {
    return _blockToolbar.default;
  }
});
Object.defineProperty(exports, "CopyHandler", {
  enumerable: true,
  get: function get() {
    return _copyHandler.default;
  }
});
Object.defineProperty(exports, "DefaultBlockAppender", {
  enumerable: true,
  get: function get() {
    return _defaultBlockAppender.default;
  }
});
Object.defineProperty(exports, "Inserter", {
  enumerable: true,
  get: function get() {
    return _inserter.default;
  }
});
Object.defineProperty(exports, "MultiBlocksSwitcher", {
  enumerable: true,
  get: function get() {
    return _multiBlocksSwitcher.default;
  }
});
Object.defineProperty(exports, "MultiSelectScrollIntoView", {
  enumerable: true,
  get: function get() {
    return _multiSelectScrollIntoView.default;
  }
});
Object.defineProperty(exports, "NavigableToolbar", {
  enumerable: true,
  get: function get() {
    return _navigableToolbar.default;
  }
});
Object.defineProperty(exports, "ObserveTyping", {
  enumerable: true,
  get: function get() {
    return _observeTyping.default;
  }
});
Object.defineProperty(exports, "PreserveScrollInReorder", {
  enumerable: true,
  get: function get() {
    return _preserveScrollInReorder.default;
  }
});
Object.defineProperty(exports, "SkipToSelectedBlock", {
  enumerable: true,
  get: function get() {
    return _skipToSelectedBlock.default;
  }
});
Object.defineProperty(exports, "Typewriter", {
  enumerable: true,
  get: function get() {
    return _typewriter.default;
  }
});
Object.defineProperty(exports, "Warning", {
  enumerable: true,
  get: function get() {
    return _warning.default;
  }
});
Object.defineProperty(exports, "WritingFlow", {
  enumerable: true,
  get: function get() {
    return _writingFlow.default;
  }
});
Object.defineProperty(exports, "BlockEditorProvider", {
  enumerable: true,
  get: function get() {
    return _provider.default;
  }
});

var _colors = require("./colors");

Object.keys(_colors).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _colors[key];
    }
  });
});

var _fontSizes = require("./font-sizes");

Object.keys(_fontSizes).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _fontSizes[key];
    }
  });
});

var _alignmentToolbar = _interopRequireDefault(require("./alignment-toolbar"));

var _autocomplete = _interopRequireDefault(require("./autocomplete"));

var _blockAlignmentToolbar = _interopRequireDefault(require("./block-alignment-toolbar"));

var _blockControls = _interopRequireDefault(require("./block-controls"));

var _blockEdit = _interopRequireDefault(require("./block-edit"));

var _blockFormatControls = _interopRequireDefault(require("./block-format-controls"));

var _blockIcon = _interopRequireDefault(require("./block-icon"));

var _dropdown = _interopRequireDefault(require("./block-navigation/dropdown"));

var _list = _interopRequireDefault(require("./block-navigation/list"));

var _blockVerticalAlignmentToolbar = _interopRequireDefault(require("./block-vertical-alignment-toolbar"));

var _buttonBlockAppender = _interopRequireDefault(require("./button-block-appender"));

var _colorPalette = _interopRequireDefault(require("./color-palette"));

var _contrastChecker = _interopRequireDefault(require("./contrast-checker"));

var _innerBlocks = _interopRequireDefault(require("./inner-blocks"));

var _inspectorAdvancedControls = _interopRequireDefault(require("./inspector-advanced-controls"));

var _inspectorControls = _interopRequireDefault(require("./inspector-controls"));

var _mediaPlaceholder = _interopRequireDefault(require("./media-placeholder"));

var _mediaUpload = _interopRequireDefault(require("./media-upload"));

var _check = _interopRequireDefault(require("./media-upload/check"));

var _panelColorSettings = _interopRequireDefault(require("./panel-color-settings"));

var _plainText = _interopRequireDefault(require("./plain-text"));

var _richText = _interopRequireWildcard(require("./rich-text"));

var _urlInput = _interopRequireDefault(require("./url-input"));

var _button = _interopRequireDefault(require("./url-input/button"));

var _urlPopover = _interopRequireDefault(require("./url-popover"));

var _withColorContext = _interopRequireDefault(require("./color-palette/with-color-context"));

var _blockSettingsMenuFirstItem = _interopRequireDefault(require("./block-settings-menu/block-settings-menu-first-item"));

var _blockSettingsMenuPluginsExtension = _interopRequireDefault(require("./block-settings-menu/block-settings-menu-plugins-extension"));

var _inserterMenuExtension = _interopRequireDefault(require("./inserter-menu-extension"));

var _blockEditorKeyboardShortcuts = _interopRequireDefault(require("./block-editor-keyboard-shortcuts"));

var _blockInspector = _interopRequireDefault(require("./block-inspector"));

var _blockList = _interopRequireDefault(require("./block-list"));

var _blockMover = _interopRequireDefault(require("./block-mover"));

var _blockPreview = _interopRequireDefault(require("./block-preview"));

var _blockSelectionClearer = _interopRequireDefault(require("./block-selection-clearer"));

var _blockSettingsMenu = _interopRequireDefault(require("./block-settings-menu"));

var _blockTitle = _interopRequireDefault(require("./block-title"));

var _blockToolbar = _interopRequireDefault(require("./block-toolbar"));

var _copyHandler = _interopRequireDefault(require("./copy-handler"));

var _defaultBlockAppender = _interopRequireDefault(require("./default-block-appender"));

var _inserter = _interopRequireDefault(require("./inserter"));

var _multiBlocksSwitcher = _interopRequireDefault(require("./block-switcher/multi-blocks-switcher"));

var _multiSelectScrollIntoView = _interopRequireDefault(require("./multi-select-scroll-into-view"));

var _navigableToolbar = _interopRequireDefault(require("./navigable-toolbar"));

var _observeTyping = _interopRequireDefault(require("./observe-typing"));

var _preserveScrollInReorder = _interopRequireDefault(require("./preserve-scroll-in-reorder"));

var _skipToSelectedBlock = _interopRequireDefault(require("./skip-to-selected-block"));

var _typewriter = _interopRequireDefault(require("./typewriter"));

var _warning = _interopRequireDefault(require("./warning"));

var _writingFlow = _interopRequireDefault(require("./writing-flow"));

var _provider = _interopRequireDefault(require("./provider"));
