// @wordpress/block-editor/build/components/block-mover/index.native.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _lodash = require("lodash");

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _data = require("@wordpress/data");

var _compose = require("@wordpress/compose");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */
var BlockMover = function BlockMover(_ref) {
  var isFirst = _ref.isFirst,
      isLast = _ref.isLast,
      onMoveDown = _ref.onMoveDown,
      onMoveUp = _ref.onMoveUp,
      firstIndex = _ref.firstIndex;
  return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_components.ToolbarButton, {
    title: !isFirst ? (0, _i18n.sprintf)(
    /* translators: accessibility text. %1: current block position (number). %2: next block position (number) */
    (0, _i18n.__)('Move block up from row %1$s to row %2$s'), firstIndex + 1, firstIndex) : (0, _i18n.__)('Move block up'),
    isDisabled: isFirst,
    onClick: onMoveUp,
    icon: "arrow-up-alt",
    extraProps: {
      hint: (0, _i18n.__)('Double tap to move the block up')
    }
  }), (0, _element.createElement)(_components.ToolbarButton, {
    title: !isLast ? (0, _i18n.sprintf)(
    /* translators: accessibility text. %1: current block position (number). %2: next block position (number) */
    (0, _i18n.__)('Move block down from row %1$s to row %2$s'), firstIndex + 1, firstIndex + 2) : (0, _i18n.__)('Move block down'),
    isDisabled: isLast,
    onClick: onMoveDown,
    icon: "arrow-down-alt",
    extraProps: {
      hint: (0, _i18n.__)('Double tap to move the block down')
    }
  }));
};

var _default = (0, _compose.compose)((0, _data.withSelect)(function (select, _ref2) {
  var clientIds = _ref2.clientIds;

  var _select = select('core/block-editor'),
      getBlockIndex = _select.getBlockIndex,
      getBlockRootClientId = _select.getBlockRootClientId,
      getBlockOrder = _select.getBlockOrder;

  var normalizedClientIds = (0, _lodash.castArray)(clientIds);
  var firstClientId = (0, _lodash.first)(normalizedClientIds);
  var rootClientId = getBlockRootClientId((0, _lodash.first)(normalizedClientIds));
  var blockOrder = getBlockOrder(rootClientId);
  var firstIndex = getBlockIndex(firstClientId, rootClientId);
  var lastIndex = getBlockIndex((0, _lodash.last)(normalizedClientIds), rootClientId);
  return {
    firstIndex: firstIndex,
    isFirst: firstIndex === 0,
    isLast: lastIndex === blockOrder.length - 1
  };
}), (0, _data.withDispatch)(function (dispatch, _ref3) {
  var clientIds = _ref3.clientIds,
      rootClientId = _ref3.rootClientId;

  var _dispatch = dispatch('core/block-editor'),
      moveBlocksDown = _dispatch.moveBlocksDown,
      moveBlocksUp = _dispatch.moveBlocksUp;

  return {
    onMoveDown: (0, _lodash.partial)(moveBlocksDown, clientIds, rootClientId),
    onMoveUp: (0, _lodash.partial)(moveBlocksUp, clientIds, rootClientId)
  };
}), _compose.withInstanceId)(BlockMover);

exports.default = _default;
