// @wordpress/editor/build/store/reducer.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  postTitle: true
};
exports.default = exports.postTitle = void 0;

var _reduxOptimist = _interopRequireDefault(require("redux-optimist"));

var _data = require("@wordpress/data");

var _reducer = require("./reducer.js");

Object.keys(_reducer).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _reducer[key];
    }
  });
});

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

/**
 * Reducer returning the post title state.
 *
 * @param {PostTitleState} state  Current state.
 * @param {Object}         action Dispatched action.
 *
 * @return {Object} Updated state.
 */
var postTitle = (0, _data.combineReducers)({
  isSelected: function isSelected() {
    var state = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;
    var action = arguments.length > 1 ? arguments[1] : undefined;

    switch (action.type) {
      case 'TOGGLE_POST_TITLE_SELECTION':
        return action.isSelected;
    }

    return state;
  }
});
exports.postTitle = postTitle;

var _default = (0, _reduxOptimist.default)((0, _data.combineReducers)({
  editor: _reducer.editor,
  initialEdits: _reducer.initialEdits,
  currentPost: _reducer.currentPost,
  preferences: _reducer.preferences,
  saving: _reducer.saving,
  postLock: _reducer.postLock,
  reusableBlocks: _reducer.reusableBlocks,
  template: _reducer.template,
  previewLink: _reducer.previewLink,
  postSavingLock: _reducer.postSavingLock,
  isReady: _reducer.isReady,
  editorSettings: _reducer.editorSettings,
  postTitle: postTitle
}));

exports.default = _default;
