// @wordpress/components/build/keyboard-shortcuts/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _mousetrap = _interopRequireDefault(require("mousetrap"));

require("mousetrap/plugins/global-bind/mousetrap-global-bind");

var _lodash = require("lodash");

var _platform = require("./platform");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var KeyboardShortcuts =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(KeyboardShortcuts, _Component);

  function KeyboardShortcuts() {
    var _this;

    (0, _classCallCheck2.default)(this, KeyboardShortcuts);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(KeyboardShortcuts).apply(this, arguments));
    _this.bindKeyTarget = _this.bindKeyTarget.bind((0, _assertThisInitialized2.default)(_this));
    return _this;
  }

  (0, _createClass2.default)(KeyboardShortcuts, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this2 = this;

      var _this$keyTarget = this.keyTarget,
          keyTarget = _this$keyTarget === void 0 ? document : _this$keyTarget;
      this.mousetrap = new _mousetrap.default(keyTarget);
      (0, _lodash.forEach)(this.props.shortcuts, function (callback, key) {
        if (process.env.NODE_ENV === 'development') {
          var keys = key.split('+');
          var modifiers = new Set(keys.filter(function (value) {
            return value.length > 1;
          }));
          var hasAlt = modifiers.has('alt');
          var hasShift = modifiers.has('shift');

          if ((0, _platform.isAppleOS)() && (modifiers.size === 1 && hasAlt || modifiers.size === 2 && hasAlt && hasShift)) {
            throw new Error("Cannot bind ".concat(key, ". Alt and Shift+Alt modifiers are reserved for character input."));
          }
        }

        var _this2$props = _this2.props,
            bindGlobal = _this2$props.bindGlobal,
            eventName = _this2$props.eventName;
        var bindFn = bindGlobal ? 'bindGlobal' : 'bind';

        _this2.mousetrap[bindFn](key, callback, eventName);
      });
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.mousetrap.reset();
    }
    /**
     * When rendering with children, binds the wrapper node on which events
     * will be bound.
     *
     * @param {Element} node Key event target.
     */

  }, {
    key: "bindKeyTarget",
    value: function bindKeyTarget(node) {
      this.keyTarget = node;
    }
  }, {
    key: "render",
    value: function render() {
      // Render as non-visual if there are no children pressed. Keyboard
      // events will be bound to the document instead.
      var children = this.props.children;

      if (!_element.Children.count(children)) {
        return null;
      }

      return (0, _element.createElement)("div", {
        ref: this.bindKeyTarget
      }, children);
    }
  }]);
  return KeyboardShortcuts;
}(_element.Component);

var _default = KeyboardShortcuts;
exports.default = _default;
