// @wordpress/compose/build/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "compose", {
  enumerable: true,
  get: function get() {
    return _lodash.flowRight;
  }
});
Object.defineProperty(exports, "createHigherOrderComponent", {
  enumerable: true,
  get: function get() {
    return _createHigherOrderComponent.default;
  }
});
Object.defineProperty(exports, "ifCondition", {
  enumerable: true,
  get: function get() {
    return _ifCondition.default;
  }
});
Object.defineProperty(exports, "pure", {
  enumerable: true,
  get: function get() {
    return _pure.default;
  }
});
Object.defineProperty(exports, "withGlobalEvents", {
  enumerable: true,
  get: function get() {
    return _withGlobalEvents.default;
  }
});
Object.defineProperty(exports, "withInstanceId", {
  enumerable: true,
  get: function get() {
    return _withInstanceId.default;
  }
});
Object.defineProperty(exports, "withSafeTimeout", {
  enumerable: true,
  get: function get() {
    return _withSafeTimeout.default;
  }
});
Object.defineProperty(exports, "withState", {
  enumerable: true,
  get: function get() {
    return _withState.default;
  }
});
Object.defineProperty(exports, "useMediaQuery", {
  enumerable: true,
  get: function get() {
    return _useMediaQuery.default;
  }
});
Object.defineProperty(exports, "useReducedMotion", {
  enumerable: true,
  get: function get() {
    return _useReducedMotion.default;
  }
});

var _lodash = require("lodash");

var _createHigherOrderComponent = _interopRequireDefault(require("./utils/create-higher-order-component"));

var _ifCondition = _interopRequireDefault(require("./higher-order/if-condition"));

var _pure = _interopRequireDefault(require("./higher-order/pure"));

var _withGlobalEvents = _interopRequireDefault(require("./higher-order/with-global-events"));

var _withInstanceId = _interopRequireDefault(require("./higher-order/with-instance-id"));

var _withSafeTimeout = _interopRequireDefault(require("./higher-order/with-safe-timeout"));

var _withState = _interopRequireDefault(require("./higher-order/with-state"));

var _useMediaQuery = _interopRequireDefault(require("./hooks/use-media-query"));

var _useReducedMotion = _interopRequireDefault(require("./hooks/use-reduced-motion"));
